
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>VirAnnot 0.0.1 documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
  </head>
  <body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="nav-item nav-item-0"><a href="index.html#document-index">VirAnnot 0.0.1 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <a class="reference internal image-reference" href="_images/INRA_logo.jpg"><img alt="_images/INRA_logo.jpg" class="align-left" src="_images/INRA_logo.jpg" style="width: 105.24px; height: 43.14px;" /></a>
<a class="reference internal image-reference" href="_images/ubx-logo.png"><img alt="_images/ubx-logo.png" class="align-right" src="_images/ubx-logo.png" style="width: 107.34px; height: 36.3px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
<div class="section" id="welcome-to-virannot-s-documentation">
<h1>Welcome to VirAnnot’s documentation!<a class="headerlink" href="#welcome-to-virannot-s-documentation" title="Permalink to this headline">¶</a></h1>
<p>VirAnnot was build to ease the assembly, blast search and taxonomic annotation of metagenomic multi-sample datasets. It is used in the Virologie team of <a class="reference external" href="http://www6.bordeaux-aquitaine.inra.fr/bfp">UMR1332 BFP</a> laboratory at INRA.</p>
<p>It was designed to identify viruses in plants but it can be used to assemble and then annotate any sequences with the NCBI taxonomy.</p>
<p>NR and NT must be present localy and/or on distant servers and NCBI taxonomy is loaded in SQLITE database with a provided script.</p>
<p>Blast step is the most time consumming step and the use of large computer cluster is clearly an advantage.
Here we used two clusters :</p>
<p><a class="reference external" href="https://redmine.mcia.univ-bordeaux.fr/">AVAKAS</a> at Bordeaux University.</p>
<p><a class="reference external" href="http://www.genotoul.fr//">GENOTOUL</a> at Toulouse INRA datacenter.</p>
<div class="section" id="pipeline-general-scheme">
<h2>Pipeline general scheme:<a class="headerlink" href="#pipeline-general-scheme" title="Permalink to this headline">¶</a></h2>
<img alt="_images/dia-intro.png" src="_images/dia-intro.png" />
</div>
</div>
<div class="section" id="guide">
<h1>Guide<a class="headerlink" href="#guide" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
<span id="document-prerequisite"></span><div class="section" id="prerequisite">
<h2>Prerequisite<a class="headerlink" href="#prerequisite" title="Permalink to this headline">¶</a></h2>
<div class="section" id="external-programs">
<h3>External programs<a class="headerlink" href="#external-programs" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>NCBI Blast+ suite (<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST">ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST</a>)</li>
<li>SQLite (<a class="reference external" href="https://www.sqlite.org/">https://www.sqlite.org/</a>)</li>
<li>Mummer3 (<a class="reference external" href="http://mummer.sourceforge.net/">http://mummer.sourceforge.net/</a>)</li>
<li>Bowtie2 (<a class="reference external" href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">http://bowtie-bio.sourceforge.net/bowtie2/index.shtml</a>)</li>
<li>Cutadapt (<a class="reference external" href="https://github.com/marcelm/cutadapt">https://github.com/marcelm/cutadapt</a>)</li>
<li>ETE tree (<a class="reference external" href="http://etetoolkit.org/">http://etetoolkit.org/</a>)</li>
<li>IDBA-UD (<a class="reference external" href="https://github.com/loneknightpy/idba">https://github.com/loneknightpy/idba</a>)</li>
<li>drVM (<a class="reference external" href="https://sourceforge.net/projects/sb2nhri/files/drVM/">https://sourceforge.net/projects/sb2nhri/files/drVM/</a>)</li>
<li>Open Grid Scheduler (<a class="reference external" href="http://gridscheduler.sourceforge.net/">http://gridscheduler.sourceforge.net/</a>)</li>
<li>Diamond (<a class="reference external" href="https://github.com/bbuchfink/diamond">https://github.com/bbuchfink/diamond</a>)</li>
</ul>
</div>
<div class="section" id="external-databases">
<h3>External databases<a class="headerlink" href="#external-databases" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>NCBI nr, nt (<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/blast/db/">ftp://ftp.ncbi.nlm.nih.gov/blast/db/</a>)</li>
<li>NCBI Taxonomy (<a class="reference external" href="ftp://ftp.ncbi.nih.gov/pub/taxonomy">ftp://ftp.ncbi.nih.gov/pub/taxonomy</a>)</li>
<li>PFAM (<a class="reference external" href="ftp://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/">ftp://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/</a>) (rpsblast files, fasta files, and smp files)</li>
</ul>
</div>
<div class="section" id="perl-external-libraries">
<h3>Perl external libraries<a class="headerlink" href="#perl-external-libraries" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>Getopt::Long</li>
<li>File::Basename</li>
<li>DBI</li>
<li>Data::Dumper</li>
<li>Bioperl</li>
<li>Color::Rgb</li>
<li>List::Util</li>
<li>Excel::Writer</li>
<li>Log::Log4perl</li>
<li>DBD::SQLite</li>
<li>SQL::SplitStatement</li>
<li>Math::Round</li>
</ul>
</div>
<div class="section" id="perl-included-libraries">
<h3>Perl included libraries<a class="headerlink" href="#perl-included-libraries" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>Tools::Fasta</li>
<li>Tools::Fastq</li>
<li>Tools::Blast</li>
<li>Tools::Taxonomy</li>
<li>Logger::Logger</li>
</ul>
</div>
<div class="section" id="python-library">
<h3>Python library<a class="headerlink" href="#python-library" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>os</li>
<li>call</li>
<li>logging</li>
<li>random</li>
<li>string</li>
<li>argparse</li>
<li>re</li>
<li>sys</li>
<li>Bio</li>
<li>time</li>
<li>glob</li>
<li>shutil</li>
<li>yaml</li>
<li>csv</li>
<li>importlib</li>
<li>matplotlib</li>
</ul>
</div>
</div>
<div class="section" id="install">
<h2>Install<a class="headerlink" href="#install" title="Permalink to this headline">¶</a></h2>
<p>Add tools and launchers folders to your $PATH.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="nb">export</span> <span class="nv">PATH</span><span class="o">=</span>/path/to/tools:/path/to/launchers:<span class="nv">$PATH</span>
</pre></div>
</div>
<p>Add lib folder to your $PERL5LIB.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="nb">export</span> <span class="nv">PERL5LIB</span><span class="o">=</span>/path/to/lib:<span class="nv">$PERL5LIB</span>
</pre></div>
</div>
</div>
<div class="section" id="database">
<h2>Database<a class="headerlink" href="#database" title="Permalink to this headline">¶</a></h2>
<p>NCBI taxonomy and the homemade per domain Pfam taxonomy are stored in a simple SQLite database.</p>
<p>Schema:</p>
<img alt="_images/taxo_sql.svg" src="_images/taxo_sql.svg" /><div class="section" id="ncbi-taxonomy">
<h3>NCBI Taxonomy<a class="headerlink" href="#ncbi-taxonomy" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>Download and extract NCBI taxonomy files.</li>
</ul>
<div class="highlight-bash"><div class="highlight"><pre><span></span>wget ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz <span class="p">;</span> gunzip taxdump.tar.gz<span class="p">;</span> tar -xf taxdump.tar<span class="p">;</span>
wget ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/prot.accession2taxid.gz <span class="p">;</span> gunzip prot.accession2taxid.gz<span class="p">;</span>
wget ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/nucl_gb.accession2taxid.gz <span class="p">;</span> gunzip nucl_gb.accession2taxid.gz<span class="p">;</span>
</pre></div>
</div>
<p>Optionally you can combine multiple accession2taxid file with a simple cat. But keep separated nucl and prot accessions as they will be loaded in two different tables.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>wget ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/dead_prot.accession2taxid.gz <span class="p">;</span> gunzip dead_prot.accession2taxid.gz<span class="p">;</span>
cat prot.accession2taxid dead_prot.accession2taxid &gt; acc2taxid.prot

wget ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/nucl_wgs.accession2taxid.gz <span class="p">;</span> gunzip nucl_wgs.accession2taxid.gz<span class="p">;</span>
wget ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/dead_wgs.accession2taxid.gz <span class="p">;</span> gunzip dead_wgs.accession2taxid.gz
cat nucl_wgs.accession2taxid nucl_gb.accession2taxid dead_wgs.accession2taxid &gt; acc2taxid.nucl
</pre></div>
</div>
<p>Launch the loadTaxonomy.pl script that will create the sqlite database. The script needs two provided sqlite files: <code class="docutils literal"><span class="pre">taxonomyIndex.sql</span></code> and <code class="docutils literal"><span class="pre">taxonomyStructure.sql</span></code> that describe the database struture.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>loadTaxonomy.pl -struct taxonomyStructure.sql -index taxonomyIndex.sql -acc_prot acc2taxid.prot -acc_nucl acc2taxid.nucl -names names.dmp -nodes nodes.dmp
</pre></div>
</div>
</div>
<div class="section" id="pfam-taxonomy">
<h3>PFAM taxonomy<a class="headerlink" href="#pfam-taxonomy" title="Permalink to this headline">¶</a></h3>
<p>The pipeline modules <code class="docutils literal"><span class="pre">rps2ecsv</span></code> and <code class="docutils literal"><span class="pre">rps2tree</span></code> need taxonomic information of the PFAM domains to work.
You need to extract these informations and load it into the sqlite database.</p>
<ul class="simple">
<li>Extract taxonomic information for each sequence of each PFAM domain and store it in <code class="docutils literal"><span class="pre">*.tax.txt</span></code> files:</li>
</ul>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="se">\l</span>s -1 *.FASTA <span class="p">|</span> sed <span class="s1">&#39;s,^\(.*\)\.FASTA,gi2taxonomy.pl -i &amp; -o \1.tax.txt -r,&#39;</span> <span class="p">|</span> bash
</pre></div>
</div>
<ul class="simple">
<li>Create a file of file for the <code class="docutils literal"><span class="pre">*.tax.txt</span></code> files:</li>
</ul>
<div class="highlight-bash"><div class="highlight"><pre><span></span>listPath.pl -d . <span class="p">|</span> grep <span class="s1">&#39;tax.txt&#39;</span> &gt; idx
</pre></div>
</div>
<ul class="simple">
<li>Compute taxonomy statistic for each domain and create a sql file to load into the database:</li>
</ul>
<div class="highlight-bash"><div class="highlight"><pre><span></span>taxo_profile_to_sql.pl -i idx -o taxo_profile.sql
</pre></div>
</div>
<ul class="simple">
<li>Load into the database:</li>
</ul>
<div class="highlight-bash"><div class="highlight"><pre><span></span>sqlite3 taxonomy.tmp.sqlite &lt; taxo_profile.sql
</pre></div>
</div>
</div>
<div class="section" id="ncbi-blast-database">
<h3>NCBI Blast database<a class="headerlink" href="#ncbi-blast-database" title="Permalink to this headline">¶</a></h3>
<p>NCBI non redundant databases are very large and similarity search using Blast is an intensive task. I recommand to use those databases on computer clusters.</p>
<ul class="simple">
<li>Download NCBI nr et nt Blast files.</li>
</ul>
<div class="highlight-bash"><div class="highlight"><pre><span></span>wget ftp://ftp.ncbi.nlm.nih.gov/blast/db/nr.*.tar.gz
wget ftp://ftp.ncbi.nlm.nih.gov/blast/db/nt.*.tar.gz
</pre></div>
</div>
<p>Modify the parameters.yaml to fit your configuration.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">servers</span><span class="p">:</span>
<span class="n">genotoul</span><span class="p">:</span>
  <span class="n">adress</span><span class="p">:</span> <span class="s1">&#39;genotoul.toulouse.inra.fr&#39;</span>
  <span class="n">username</span><span class="p">:</span> <span class="s1">&#39;stheil&#39;</span>
  <span class="n">db</span><span class="p">:</span>
    <span class="n">nr</span><span class="p">:</span> <span class="s1">&#39;/bank/blastdb/nr&#39;</span>
    <span class="n">nt</span><span class="p">:</span> <span class="s1">&#39;/bank/blastdb/nt&#39;</span>
</pre></div>
</div>
<p>Reduced databases are a good choice for limited computer ressources and drastically faster similarity search. Here are some example commands using NCBI tools to download sequences.</p>
<ul class="simple">
<li>Reduced NCBI databases:</li>
</ul>
<p>Get all viroids nucleotide sequence from genbank:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">esearch</span> <span class="o">-</span><span class="n">db</span> <span class="s2">&quot;nucleotide&quot;</span> <span class="o">-</span><span class="n">query</span> <span class="s2">&quot;txid12884[Organism]&quot;</span> <span class="o">|</span> <span class="n">efetch</span> <span class="o">-</span><span class="nb">format</span> <span class="n">fasta</span> <span class="o">&gt;</span> <span class="n">viroids_nucl</span><span class="o">.</span><span class="n">fna</span>
</pre></div>
</div>
<p>Get all viruses nucleotide sequences from genbank:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">esearch</span> <span class="o">-</span><span class="n">db</span> <span class="s2">&quot;nucleotide&quot;</span> <span class="o">-</span><span class="n">query</span> <span class="s2">&quot;txid10239[Organism]&quot;</span> <span class="o">|</span> <span class="n">efetch</span> <span class="o">-</span><span class="nb">format</span> <span class="n">fasta</span> <span class="o">&gt;</span> <span class="n">viruses_nucl</span><span class="o">.</span><span class="n">fna</span>
</pre></div>
</div>
<p>Create Blast DB example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">makeblastdb</span> <span class="o">-</span><span class="ow">in</span> <span class="n">viruses_nucl</span><span class="o">.</span><span class="n">fna</span> <span class="o">-</span><span class="n">parse_seqids</span> <span class="o">-</span><span class="n">dbtype</span> <span class="n">nucl</span>
</pre></div>
</div>
<ul class="simple">
<li>Download PFAM files for RPSBLAST.</li>
</ul>
<div class="highlight-bash"><div class="highlight"><pre><span></span>wget ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/little_endian/Pfam_LE.tar.gz
wget ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/fasta.tar.gz
wget ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/cdd.tar.gz
</pre></div>
</div>
<p>Here I use only PFAM domains but <code class="docutils literal"><span class="pre">fasta.tar.gz</span></code> and <code class="docutils literal"><span class="pre">cdd.tar.gz</span></code> contains files for the entire CDD database. You can either delete files that are not from PFAM database or use the complete CDD.</p>
<ul class="simple">
<li>Delete file that are not from PFAM:</li>
</ul>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="se">\l</span>s -1 <span class="p">|</span> grep -v <span class="s1">&#39;pfam&#39;</span> <span class="p">|</span> sed <span class="s1">&#39;s,^.*$,rm &amp;,&#39;</span>
</pre></div>
</div>
<p>Add ‘| bash’ if correct.</p>
<ul class="simple">
<li>Download entire CDD database:</li>
</ul>
<div class="highlight-bash"><div class="highlight"><pre><span></span>wget ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/little_endian/CDD_LE.tar.gz
</pre></div>
</div>
</div>
</div>
<span id="document-parameter"></span><div class="section" id="parameters-files">
<h2>Parameters files<a class="headerlink" href="#parameters-files" title="Permalink to this headline">¶</a></h2>
<div class="section" id="parameters-yaml">
<h3>parameters.yaml<a class="headerlink" href="#parameters-yaml" title="Permalink to this headline">¶</a></h3>
<p>Defines paths for both local and remote binaries and databases. A template is provided in the <code class="docutils literal"><span class="pre">examples</span></code> directory.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">ReadSoustraction</span><span class="p">:</span>
  <span class="n">db</span><span class="p">:</span>
    <span class="n">vitis</span><span class="p">:</span> <span class="s1">&#39;/media/data/db/ncbi/vitis/vitis&#39;</span>
    <span class="n">phiX</span><span class="p">:</span>  <span class="s1">&#39;/media/data/db/ncbi/phiX/phiX174&#39;</span>
<span class="nb">bin</span><span class="p">:</span>
  <span class="n">bowtie</span><span class="p">:</span> <span class="s1">&#39;/usr/local/bin/bowtie2&#39;</span>
  <span class="n">samtools</span><span class="p">:</span> <span class="s1">&#39;/usr/bin/samtools&#39;</span>
  <span class="n">bedtools</span><span class="p">:</span> <span class="s1">&#39;/usr/bin/bedtools&#39;</span>
  <span class="n">prinseq</span><span class="p">:</span> <span class="s1">&#39;/usr/local/bin/prinseq-lite.pl&#39;</span>
  <span class="n">merge</span><span class="o">-</span><span class="n">paired</span><span class="o">-</span><span class="n">reads</span><span class="p">:</span> <span class="s1">&#39;/home/stheil/softwares/sortmerna-2.1-linux-64/scripts/merge-paired-reads.sh&#39;</span>
  <span class="n">unmerge</span><span class="o">-</span><span class="n">paired</span><span class="o">-</span><span class="n">reads</span><span class="p">:</span> <span class="s1">&#39;/home/stheil/softwares/sortmerna-2.1-linux-64/scripts/unmerge-paired-reads.sh&#39;</span>
  <span class="n">sortmerna</span><span class="p">:</span> <span class="s1">&#39;/home/stheil/softwares/sortmerna/sortmerna&#39;</span>
<span class="n">servers</span><span class="p">:</span>
  <span class="n">enki</span><span class="p">:</span>
    <span class="n">db</span><span class="p">:</span>
      <span class="n">nt</span><span class="p">:</span> <span class="s1">&#39;/media/data/db/ncbi/nt/nt&#39;</span>
      <span class="n">nr</span><span class="p">:</span> <span class="s1">&#39;/media/data/db/ncbi/nr/nr&#39;</span>
      <span class="n">refseq_vir_nucl</span><span class="p">:</span> <span class="s1">&#39;/media/data/db/ncbi/refseq_vir/viral.genomic.fna&#39;</span>
      <span class="n">refseq_vir_prot</span><span class="p">:</span> <span class="s1">&#39;/media/data/db/ncbi/refseq_vir/viral.protein.faa&#39;</span>
      <span class="n">pfam</span><span class="p">:</span> <span class="s1">&#39;/home/stheil/save/db/pfam/pfam_viruses_rpsdb&#39;</span>
      <span class="n">all_vir_nucl</span><span class="p">:</span> <span class="s1">&#39;/media/data/db/ncbi/all_vir/all_vir_nucl.fna&#39;</span>
      <span class="n">all_vir_prot</span><span class="p">:</span> <span class="s1">&#39;/media/data/db/ncbi/all_vir/all_vir_prot.faa&#39;</span>
  <span class="n">genotoul</span><span class="p">:</span>
    <span class="n">adress</span><span class="p">:</span> <span class="s1">&#39;genotoul.toulouse.inra.fr&#39;</span>
    <span class="n">username</span><span class="p">:</span> <span class="s1">&#39;stheil&#39;</span>
    <span class="n">db</span><span class="p">:</span>
      <span class="n">nr</span><span class="p">:</span> <span class="s1">&#39;/bank/blastdb/nr&#39;</span>
      <span class="n">nt</span><span class="p">:</span> <span class="s1">&#39;/bank/blastdb/nt&#39;</span>
      <span class="n">refseq_vir_nucl</span><span class="p">:</span> <span class="s1">&#39;/save/stheil/db/refseq_vir/viral.genomic.fna&#39;</span>
      <span class="n">refseq_vir_prot</span><span class="p">:</span> <span class="s1">&#39;/save/stheil/db/refseq_vir/viral.protein.faa&#39;</span>
      <span class="n">pfam</span><span class="p">:</span> <span class="s1">&#39;/home/stheil/save/db/pfam/pfam_viruses_rpsdb&#39;</span>
      <span class="n">all_vir_nucl</span><span class="p">:</span> <span class="s1">&#39;/home/stheil/save/db/all_vir/all_vir_nucl.fna&#39;</span>
      <span class="n">all_vir_prot</span><span class="p">:</span> <span class="s1">&#39;/home/stheil/save/db/all_vir/all_vir_prot.faa&#39;</span>
    <span class="n">scratch</span><span class="p">:</span> <span class="s1">&#39;/work/stheil&#39;</span>
    <span class="nb">bin</span><span class="p">:</span>
      <span class="n">blastx</span><span class="p">:</span> <span class="s1">&#39;blastx+&#39;</span>
      <span class="n">blastn</span><span class="p">:</span> <span class="s1">&#39;blastn+&#39;</span>
  <span class="n">avakas</span><span class="p">:</span>
    <span class="n">adress</span><span class="p">:</span> <span class="s1">&#39;avakas.mcia.univ-bordeaux.fr&#39;</span>
    <span class="n">username</span><span class="p">:</span> <span class="s1">&#39;stheil&#39;</span>
    <span class="n">db</span><span class="p">:</span>
      <span class="n">nr</span><span class="p">:</span> <span class="s1">&#39;/home/stheil/db/nr/nr&#39;</span>
      <span class="n">nt</span><span class="p">:</span> <span class="s1">&#39;/home/stheil/db/nt/nt&#39;</span>
      <span class="n">all_vir_nucl</span><span class="p">:</span> <span class="s1">&#39;/home/stheil/scratch/db/all_vir/all_vir_nucl.fna&#39;</span>
      <span class="n">all_vir_prot</span><span class="p">:</span> <span class="s1">&#39;/home/stheil/scratch/db/all_vir/all_vir_prot.faa&#39;</span>
      <span class="n">refseq_vir_nucl</span><span class="p">:</span> <span class="s1">&#39;/home/stheil/scratch/db/refseq_vir/viral.genomic.fna&#39;</span>
      <span class="n">refseq_vir_prot</span><span class="p">:</span> <span class="s1">&#39;/home/stheil/scratch/db/refseq_vir/viral.protein.faa&#39;</span>
      <span class="n">pfam</span><span class="p">:</span> <span class="s1">&#39;/home/stheil/db/pfam/pfam_viruses_rpsdb&#39;</span>
    <span class="n">scratch</span><span class="p">:</span> <span class="s1">&#39;/scratch/stheil&#39;</span>
    <span class="nb">bin</span><span class="p">:</span>
      <span class="n">blastx</span><span class="p">:</span> <span class="s1">&#39;blastx&#39;</span>
      <span class="n">blastn</span><span class="p">:</span> <span class="s1">&#39;blastn&#39;</span>
<span class="n">Diamond</span><span class="p">:</span>
  <span class="n">db</span><span class="p">:</span>
    <span class="n">all_vir_prot</span><span class="p">:</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">db</span><span class="o">/</span><span class="n">ncbi</span><span class="o">/</span><span class="n">all_vir</span><span class="o">/</span><span class="n">all_vir_prot</span>
<span class="n">SortMeRna</span><span class="p">:</span>
  <span class="n">db</span><span class="p">:</span>
    <span class="n">silva</span><span class="o">-</span><span class="n">arc</span><span class="o">-</span><span class="mi">16</span><span class="n">s</span><span class="o">-</span><span class="n">id95</span><span class="p">:</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">db</span><span class="o">/</span><span class="n">rRNA_databases</span><span class="o">/</span><span class="n">silva</span><span class="o">-</span><span class="n">arc</span><span class="o">-</span><span class="mi">16</span><span class="n">s</span><span class="o">-</span><span class="n">id95</span>
    <span class="n">silva</span><span class="o">-</span><span class="n">arc</span><span class="o">-</span><span class="mi">23</span><span class="n">s</span><span class="o">-</span><span class="n">id98</span><span class="p">:</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">db</span><span class="o">/</span><span class="n">rRNA_databases</span><span class="o">/</span><span class="n">silva</span><span class="o">-</span><span class="n">arc</span><span class="o">-</span><span class="mi">23</span><span class="n">s</span><span class="o">-</span><span class="n">id98</span>
    <span class="n">silva</span><span class="o">-</span><span class="n">bac</span><span class="o">-</span><span class="mi">16</span><span class="n">s</span><span class="o">-</span><span class="n">id90</span><span class="p">:</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">db</span><span class="o">/</span><span class="n">rRNA_databases</span><span class="o">/</span><span class="n">silva</span><span class="o">-</span><span class="n">bac</span><span class="o">-</span><span class="mi">16</span><span class="n">s</span><span class="o">-</span><span class="n">id90</span>
    <span class="n">silva</span><span class="o">-</span><span class="n">bac</span><span class="o">-</span><span class="mi">23</span><span class="n">s</span><span class="o">-</span><span class="n">id98</span><span class="p">:</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">db</span><span class="o">/</span><span class="n">rRNA_databases</span><span class="o">/</span><span class="n">silva</span><span class="o">-</span><span class="n">bac</span><span class="o">-</span><span class="mi">23</span><span class="n">s</span><span class="o">-</span><span class="n">id98</span>
    <span class="n">silva</span><span class="o">-</span><span class="n">euk</span><span class="o">-</span><span class="mi">18</span><span class="n">s</span><span class="o">-</span><span class="n">id95</span><span class="p">:</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">db</span><span class="o">/</span><span class="n">rRNA_databases</span><span class="o">/</span><span class="n">silva</span><span class="o">-</span><span class="n">euk</span><span class="o">-</span><span class="mi">18</span><span class="n">s</span><span class="o">-</span><span class="n">id95</span>
    <span class="n">silva</span><span class="o">-</span><span class="n">euk</span><span class="o">-</span><span class="mi">28</span><span class="n">s</span><span class="o">-</span><span class="n">id98</span><span class="p">:</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">db</span><span class="o">/</span><span class="n">rRNA_databases</span><span class="o">/</span><span class="n">silva</span><span class="o">-</span><span class="n">euk</span><span class="o">-</span><span class="mi">28</span><span class="n">s</span><span class="o">-</span><span class="n">id98</span>
</pre></div>
</div>
</div>
<div class="section" id="step-yaml">
<h3>step.yaml<a class="headerlink" href="#step-yaml" title="Permalink to this headline">¶</a></h3>
<p>Defines the steps that the pipeline will execute. A template is provided in the <code class="docutils literal"><span class="pre">/examples</span></code> directory.</p>
<p>Step names correspond to a python module that will launch the step. Step names are split based on the ‘_’ character so you can launch multiple instance. For example you might want to launch blastx and blastn, so step names could be ‘Blast_N’ and ‘Blast_X’. What is after the underscore do not matters, it is just used to differanciate the two steps.</p>
<p>Special words in bracket are used as substitution string.
- (file), (file1) and (file2)
- (SampleID)
- (library)</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">ReadSoustraction_phiX</span><span class="p">:</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">file1</span><span class="p">)</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">file2</span><span class="p">)</span>
  <span class="n">db</span><span class="p">:</span> <span class="n">phiX</span>
  <span class="n">o1</span><span class="p">:</span> <span class="p">(</span><span class="n">library</span><span class="p">)</span><span class="n">_phiX</span><span class="o">.</span><span class="n">r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">o2</span><span class="p">:</span> <span class="p">(</span><span class="n">library</span><span class="p">)</span><span class="n">_phiX</span><span class="o">.</span><span class="n">r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">5</span>
  <span class="nb">iter</span><span class="p">:</span> <span class="n">library</span>
<span class="n">Demultiplex</span><span class="p">:</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">library</span><span class="p">)</span><span class="n">_phiX</span><span class="o">.</span><span class="n">r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">library</span><span class="p">)</span><span class="n">_phiX</span><span class="o">.</span><span class="n">r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">adapters</span><span class="p">:</span> <span class="n">adapters</span><span class="o">.</span><span class="n">fna</span>
  <span class="n">middle</span><span class="p">:</span> <span class="mi">1</span>
  <span class="n">min_qual</span><span class="p">:</span> <span class="mi">20</span>
  <span class="n">polyA</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">min_len</span><span class="p">:</span> <span class="mi">70</span>
  <span class="nb">iter</span><span class="p">:</span> <span class="n">library</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">DemultiplexHtml</span><span class="p">:</span>
  <span class="n">csv</span><span class="p">:</span> <span class="p">(</span><span class="n">library</span><span class="p">)</span><span class="n">_demultiplex</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">csv</span>
  <span class="nb">id</span><span class="p">:</span> <span class="p">(</span><span class="n">library</span><span class="p">)</span>
  <span class="n">out</span><span class="p">:</span> <span class="n">stat_demultiplex</span>
  <span class="nb">iter</span><span class="p">:</span> <span class="k">global</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Normalization</span><span class="p">:</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">o1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_norm_r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">o2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_norm_r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">num</span><span class="p">:</span> <span class="mi">40000</span>
  <span class="nb">iter</span><span class="p">:</span> <span class="n">sample</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">5</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">drVM</span><span class="p">:</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">20</span>
  <span class="n">identity</span><span class="p">:</span> <span class="mi">70</span>
  <span class="n">min_len</span><span class="p">:</span> <span class="mi">300</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Assembly_idba</span><span class="p">:</span>
  <span class="n">prog</span><span class="p">:</span> <span class="n">idba</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">5</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Assembly_spades</span><span class="p">:</span>
  <span class="n">prog</span><span class="p">:</span> <span class="n">spades</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">5</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_spades</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Map_idba</span><span class="p">:</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">bam</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">bam</span>
  <span class="n">rn</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">rn</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">16</span>
<span class="n">Map_spades</span><span class="p">:</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_spades</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">bam</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_spades</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">bam</span>
  <span class="n">rn</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_spades</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">rn</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">16</span>
<span class="n">Diamond</span><span class="p">:</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">10</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">score</span><span class="p">:</span> <span class="mi">50</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.0001</span>
  <span class="n">qov</span><span class="p">:</span> <span class="mi">50</span>
  <span class="n">hov</span><span class="p">:</span> <span class="mi">5</span>
  <span class="n">db</span><span class="p">:</span> <span class="n">all_vir_prot</span>
<span class="n">Diamond_singletons_nr</span><span class="p">:</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">db</span><span class="p">:</span> <span class="n">nr</span>
  <span class="n">ising</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_singletons</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">10</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_singletons_test</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">dmdx</span><span class="o">.</span><span class="n">xml</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.001</span>
  <span class="nb">iter</span><span class="p">:</span> <span class="n">sample</span>
  <span class="n">score</span><span class="p">:</span> <span class="mi">10</span>
  <span class="n">qov</span><span class="p">:</span> <span class="mi">10</span>
<span class="n">Diamond2blast</span><span class="p">:</span>
  <span class="n">i</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">dmdx</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">dmdx2bltx</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">dmdx2bltx</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">xml</span>
  <span class="nb">type</span><span class="p">:</span> <span class="n">blastx</span>
  <span class="n">db</span><span class="p">:</span> <span class="n">nr</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.0001</span>
  <span class="n">server</span><span class="p">:</span> <span class="n">genologin</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">8</span>
  <span class="n">tc</span><span class="p">:</span> <span class="mi">50</span>
  <span class="n">num_chunk</span><span class="p">:</span> <span class="mi">1000</span>
  <span class="n">max_target_seqs</span><span class="p">:</span> <span class="mi">1</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Blast_allvirTX</span><span class="p">:</span>
  <span class="nb">type</span><span class="p">:</span> <span class="n">tblastx</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">db</span><span class="p">:</span> <span class="n">all_vir_nucl</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">all_vir</span><span class="o">.</span><span class="n">xml</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.0001</span>
  <span class="n">server</span><span class="p">:</span> <span class="n">genotoul</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">8</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">num_chunk</span><span class="p">:</span> <span class="mi">1000</span>
  <span class="n">tc</span><span class="p">:</span> <span class="mi">50</span>
<span class="n">Blast_nr</span><span class="p">:</span>
  <span class="nb">type</span><span class="p">:</span> <span class="n">blastx</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">db</span><span class="p">:</span> <span class="n">nr</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">bltx</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">xml</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.0001</span>
  <span class="n">server</span><span class="p">:</span> <span class="n">genotoul</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">8</span>
  <span class="n">tc</span><span class="p">:</span> <span class="mi">50</span>
  <span class="n">num_chunk</span><span class="p">:</span> <span class="mi">1000</span>
  <span class="n">max_target_seqs</span><span class="p">:</span> <span class="mi">1</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Blast_refvirTX</span><span class="p">:</span>
  <span class="nb">type</span><span class="p">:</span> <span class="n">tblastx</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">db</span><span class="p">:</span> <span class="n">refseq_vir_nucl</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">refseq_vir</span><span class="o">.</span><span class="n">xml</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.0001</span>
  <span class="n">server</span><span class="p">:</span> <span class="n">genotoul</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">8</span>
  <span class="n">tc</span><span class="p">:</span> <span class="mi">50</span>
  <span class="n">num_chunk</span><span class="p">:</span> <span class="mi">1000</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Blast_singleton_nr</span><span class="p">:</span>
  <span class="nb">type</span><span class="p">:</span> <span class="n">blastx</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_singletons</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">db</span><span class="p">:</span> <span class="n">nr</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_singletons</span><span class="o">.</span><span class="n">bltx</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">xml</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.0001</span>
  <span class="n">server</span><span class="p">:</span> <span class="n">genologin</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">8</span>
  <span class="n">tc</span><span class="p">:</span> <span class="mi">10</span>
  <span class="n">num_chunk</span><span class="p">:</span> <span class="mi">1000</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Blast_RPS</span><span class="p">:</span>
  <span class="nb">type</span><span class="p">:</span> <span class="n">rpstblastn</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">db</span><span class="p">:</span> <span class="n">pfam</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.0001</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">rps</span><span class="o">.</span><span class="n">pfam</span><span class="o">.</span><span class="n">xml</span>
  <span class="n">server</span><span class="p">:</span> <span class="n">genotoul</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">8</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Blast2ecsv_allvirTX</span><span class="p">:</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.001</span>
  <span class="n">fhit</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">pm</span><span class="p">:</span> <span class="k">global</span>
  <span class="k">if</span><span class="p">:</span> <span class="n">xml</span>
  <span class="n">rn</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">rn</span>
  <span class="n">r</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">b</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">all_vir</span><span class="o">.</span><span class="n">xml</span>
  <span class="n">vs</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">all_vir</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="nb">type</span><span class="p">:</span> <span class="n">TBLASTX</span>
  <span class="n">score</span><span class="p">:</span> <span class="mi">50</span>
  <span class="n">qov</span><span class="p">:</span> <span class="mi">20</span>
<span class="n">Blast2ecsv_refvirTX</span><span class="p">:</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.0001</span>
  <span class="n">fhit</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">pm</span><span class="p">:</span> <span class="k">global</span>
  <span class="k">if</span><span class="p">:</span> <span class="n">xml</span>
  <span class="n">rn</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">rn</span>
  <span class="n">r</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">b</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">refseq_vir</span><span class="o">.</span><span class="n">xml</span>
  <span class="n">vs</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">refseq_vir</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="nb">type</span><span class="p">:</span> <span class="n">TBLASTX</span>
  <span class="n">score</span><span class="p">:</span> <span class="mi">50</span>
  <span class="n">qov</span><span class="p">:</span> <span class="mi">50</span>
  <span class="n">hov</span><span class="p">:</span> <span class="mi">5</span>
<span class="n">Blast2ecsv_nr</span><span class="p">:</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.001</span>
  <span class="n">fhit</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">pm</span><span class="p">:</span> <span class="k">global</span>
  <span class="k">if</span><span class="p">:</span> <span class="n">xml</span>
  <span class="n">rn</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">rn</span>
  <span class="n">r</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">b</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">bltx</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">xml</span>
  <span class="n">vs</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">bltx</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="nb">type</span><span class="p">:</span> <span class="n">BLASTX</span>
  <span class="n">score</span><span class="p">:</span> <span class="mi">50</span>
  <span class="n">qov</span><span class="p">:</span> <span class="mi">5</span>
  <span class="n">hov</span><span class="p">:</span> <span class="mi">5</span>
<span class="n">Blast2ecsv_dmd</span><span class="p">:</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.01</span>
  <span class="n">fhit</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">pm</span><span class="p">:</span> <span class="k">global</span>
  <span class="k">if</span><span class="p">:</span> <span class="n">xml</span>
  <span class="n">r</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">b</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_dmd</span><span class="o">.</span><span class="n">xml</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_dmd</span><span class="o">.</span><span class="n">allVirProt</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="nb">type</span><span class="p">:</span> <span class="n">BLASTX</span>
  <span class="n">pd</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Blast2ecsv_dmdx_singletons_nr</span><span class="p">:</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.001</span>
  <span class="n">fhit</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">pm</span><span class="p">:</span> <span class="k">global</span>
  <span class="k">if</span><span class="p">:</span> <span class="n">xml</span>
  <span class="n">rn</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">rn</span>
  <span class="n">r</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">b</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_singletons</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">dmdx</span><span class="o">.</span><span class="n">xml</span>
  <span class="n">vs</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_singletons_test</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">dmdx</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="nb">type</span><span class="p">:</span> <span class="n">DIAMONDX</span>
  <span class="n">pd</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Rps2ecsv</span><span class="p">:</span>
  <span class="n">b</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">rps</span><span class="o">.</span><span class="n">pfam</span><span class="o">.</span><span class="n">xml</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">rps</span><span class="o">.</span><span class="n">pfam</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.0001</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Ecsv2excel</span><span class="p">:</span>
  <span class="n">b1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">refseq_vir</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">b2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">all_vir</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">b3</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">bltx</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">r</span><span class="p">:</span>  <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">rps</span><span class="o">.</span><span class="n">pfam</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">out</span><span class="p">:</span>  <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">xlsx</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Ecsv2compare</span><span class="p">:</span>
  <span class="n">b1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">bltx</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">r</span><span class="p">:</span>  <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">rps</span><span class="o">.</span><span class="n">pfam</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">out</span><span class="p">:</span>  <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">comparison</span><span class="o">.</span><span class="n">xlsx</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Blast2hist</span><span class="p">:</span>
  <span class="n">id1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_refseq_tbltx</span>
  <span class="n">b1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">refseq_vir</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">id2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_allvir_tbltx</span>
  <span class="n">b2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">all_vir</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">id3</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_nr_bltx</span>
  <span class="n">b3</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">bltx</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">id4</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_dmd</span>
  <span class="n">b4</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_dmd</span><span class="o">.</span><span class="n">allVirProt</span><span class="o">.</span><span class="n">csv</span>
  <span class="nb">iter</span><span class="p">:</span> <span class="k">global</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">out</span><span class="p">:</span> <span class="n">blast_hist</span>
<span class="n">Ecsv2krona</span><span class="p">:</span>
  <span class="n">id1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_refseq_tbltx</span>
  <span class="n">b1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">refseq_vir</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">x1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">refseq_vir</span><span class="o">.</span><span class="n">xml</span>
  <span class="n">id2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_allvir_tbltx</span>
  <span class="n">b2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">all_vir</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">x2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">all_vir</span><span class="o">.</span><span class="n">xml</span>
  <span class="n">id3</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_nr_bltx</span>
  <span class="n">b3</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">bltx</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">x3</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">bltx</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">xml</span>
  <span class="n">outdir</span><span class="p">:</span> <span class="n">krona_blast</span>
  <span class="n">out</span><span class="p">:</span> <span class="n">blast</span><span class="o">.</span><span class="k">global</span><span class="o">.</span><span class="n">krona</span><span class="o">.</span><span class="n">html</span>
  <span class="n">data</span><span class="p">:</span> <span class="n">both</span>
  <span class="n">r</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">c</span><span class="p">:</span> <span class="n">identity</span>
  <span class="nb">iter</span><span class="p">:</span> <span class="k">global</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Ecsv2krona_dmd</span><span class="p">:</span>
  <span class="n">id1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span>
  <span class="n">b1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_dmd</span><span class="o">.</span><span class="n">allVirProt</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">outdir</span><span class="p">:</span> <span class="n">krona_diamond</span>
  <span class="n">out</span><span class="p">:</span> <span class="n">global_krona_dmd</span><span class="o">.</span><span class="n">html</span>
  <span class="n">data</span><span class="p">:</span> <span class="n">contig</span>
  <span class="n">r</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">c</span><span class="p">:</span> <span class="n">identity</span>
  <span class="nb">iter</span><span class="p">:</span> <span class="k">global</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Automapper_nr</span><span class="p">:</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">ecsv</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">bltx</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_autoMapper_nr</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">ref</span><span class="p">:</span> <span class="n">nt</span>
<span class="n">Automapper_allvirTX</span><span class="p">:</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">ecsv</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">all_vir</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_autoMapper_allvir</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">ref</span><span class="p">:</span> <span class="n">all_vir_nucl</span>
<span class="n">Automapper_refseqTX</span><span class="p">:</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">ecsv</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">refseq_vir</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_autoMapper_refseq</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">ref</span><span class="p">:</span> <span class="n">refseq_vir_nucl</span>
<span class="n">Rps2tree</span><span class="p">:</span>
  <span class="n">pfam</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">rps</span><span class="o">.</span><span class="n">pfam</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">ecsv</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">bltx</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">csv</span>
  <span class="nb">id</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span>
  <span class="n">out</span><span class="p">:</span> <span class="n">rps2tree_global</span>
  <span class="n">min_prot</span><span class="p">:</span> <span class="mi">100</span>
  <span class="n">viral_portion</span><span class="p">:</span> <span class="mf">0.3</span>
  <span class="nb">iter</span><span class="p">:</span> <span class="k">global</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Getresults</span><span class="p">:</span>
  <span class="n">global_dir1</span><span class="p">:</span> <span class="n">rps2tree_global</span>
  <span class="n">global_dir2</span><span class="p">:</span> <span class="n">krona_blast</span>
  <span class="n">global_dir3</span><span class="p">:</span> <span class="n">krona_diamond</span>
  <span class="n">global_dir4</span><span class="p">:</span> <span class="n">blast_hist</span>
  <span class="n">global_dir5</span><span class="p">:</span> <span class="n">stat_demultiplex</span>
  <span class="n">sample_dir1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_autoMapper_nr</span>
  <span class="n">sample_dir2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_autoMapper_refseq</span>
  <span class="n">sample_dir3r</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_autoMapper_allvir</span>
  <span class="n">sample_file1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">xlsx</span>
  <span class="n">sample_file2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">sample_file3</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_spades</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">sample_file4</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">sample_file5</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">out</span><span class="p">:</span> <span class="n">results</span>
</pre></div>
</div>
</div>
<div class="section" id="map-txt">
<span id="id1"></span><h3>map.txt<a class="headerlink" href="#map-txt" title="Permalink to this headline">¶</a></h3>
<p>The map file describe the experiment. It is a tabulated file with the first line containing headers starting with ‘#’. It must contain at least two column: SampleID and file.
A template is provided in the <code class="docutils literal"><span class="pre">examples</span></code> directory.
This is a minimum map.txt file:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="c1">#SampleID	mid	common	file1	file2	library</span>
<span class="n">ds2016</span><span class="o">-</span><span class="mi">121</span>	<span class="n">AACCGCAA</span>	<span class="n">TGTGTTGGGTGTGTTTGG</span>	<span class="n">Lib1_phiX</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">Lib1_phiX</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">lib1</span>
<span class="n">ds2016</span><span class="o">-</span><span class="mi">132</span>	<span class="n">AACTAGTA</span>	<span class="n">TGTGTTGGGTGTGTTTGG</span>	<span class="n">Lib1_phiX</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">Lib1_phiX</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">lib1</span>
<span class="n">ds2016</span><span class="o">-</span><span class="mi">122</span>	<span class="n">AGGCGCCT</span>	<span class="n">TGTGTTGGGTGTGTTTGG</span>	<span class="n">Lib2_phiX</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">Lib2_phiX</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">lib2</span>
<span class="n">ds2016</span><span class="o">-</span><span class="mi">133</span>	<span class="n">ATTAGCTA</span>	<span class="n">TGTGTTGGGTGTGTTTGG</span>	<span class="n">Lib2_phiX</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">Lib2_phiX</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">lib2</span>
<span class="n">ds2016</span><span class="o">-</span><span class="mi">123</span>	<span class="n">CAAGAGTT</span>	<span class="n">TGTGTTGGGTGTGTTTGG</span>	<span class="n">Lib3_phiX</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">Lib3_phiX</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">lib3</span>
<span class="n">ds2016</span><span class="o">-</span><span class="mi">55</span>	<span class="n">CAAGCAGG</span>	<span class="n">TGTGTTGGGTGTGTTTGG</span>	<span class="n">Lib3_phiX</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">Lib3_phiX</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">lib3</span>
<span class="n">ds2016</span><span class="o">-</span><span class="mi">124</span>	<span class="n">CCAACCAT</span>	<span class="n">TGTGTTGGGTGTGTTTGG</span>	<span class="n">Lib4_phiX</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">Lib4_phiX</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">lib4</span>
<span class="n">ds2016</span><span class="o">-</span><span class="mi">56</span>	<span class="n">CGATAGAG</span>	<span class="n">TGTGTTGGGTGTGTTTGG</span>	<span class="n">Lib4_phiX</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">Lib4_phiX</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">lib4</span>
<span class="n">ds2016</span><span class="o">-</span><span class="mi">125</span>	<span class="n">GCTCTACC</span>	<span class="n">TGTGTTGGGTGTGTTTGG</span>	<span class="n">Lib5_phiX</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">Lib5_phiX</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">lib5</span>
<span class="n">ds2016</span><span class="o">-</span><span class="mi">57</span>	<span class="n">GCTGCGGT</span>	<span class="n">TGTGTTGGGTGTGTTTGG</span>	<span class="n">Lib5_phiX</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">Lib5_phiX</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">lib5</span>
<span class="n">ds2016</span><span class="o">-</span><span class="mi">58</span>	<span class="n">GGCCAGAA</span>	<span class="n">TGTGTTGGGTGTGTTTGG</span>	<span class="n">Lib6_phiX</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">Lib6_phiX</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">lib6</span>
<span class="n">ds2016</span><span class="o">-</span><span class="mi">10</span>	<span class="n">GGTACTCC</span>	<span class="n">TGTGTTGGGTGTGTTTGG</span>	<span class="n">Lib6_phiX</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">Lib6_phiX</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">lib6</span>
<span class="n">ds2016</span><span class="o">-</span><span class="mi">11</span>	<span class="n">TCGGATGC</span>	<span class="n">TGTGTTGGGTGTGTTTGG</span>	<span class="n">Lib7_phiX</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">Lib7_phiX</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">lib7</span>
<span class="n">ds2015</span><span class="o">-</span><span class="mi">149</span>	<span class="n">TCTATGAC</span>	<span class="n">TGTGTTGGGTGTGTTTGG</span>	<span class="n">Lib7_phiX</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">Lib7_phiX</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">lib7</span>
<span class="n">ds2015</span><span class="o">-</span><span class="mi">162</span>	<span class="n">TTCTGGCT</span>	<span class="n">TGTGTTGGGTGTGTTTGG</span>	<span class="n">Lib8_phiX</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">Lib8_phiX</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">lib8</span>
<span class="n">ds2015</span><span class="o">-</span><span class="mi">170</span>	<span class="n">TTGCGTCA</span>	<span class="n">TGTGTTGGGTGTGTTTGG</span>	<span class="n">Lib8_phiX</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">Lib8_phiX</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span>	<span class="n">lib8</span>
</pre></div>
</div>
<p>You can add categories for each sample so they can be used when coloring sequences in trees from the Rps2tree module.</p>
</div>
</div>
<span id="document-modules"></span><div class="section" id="modules-description">
<h2>Modules description<a class="headerlink" href="#modules-description" title="Permalink to this headline">¶</a></h2>
<div class="section" id="readsoustraction">
<h3>ReadSoustraction<a class="headerlink" href="#readsoustraction" title="Permalink to this headline">¶</a></h3>
<p>ReadSoustraction module use bowtie2 to map reads against a reference. The output is directly piped to samtools for bam convertion and again piped to samtools to select unmapped paired reads (-f 12 -F 256). The pipe continues with bamtofastq tool to create two fastq files (r1 and r2).</p>
<div class="highlight-python3"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">_create_cmd</span> <span class="p">(</span><span class="bp">self</span><span class="p">):</span>
  <span class="n">cmd</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
  <span class="n">cmd</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;bin&#39;</span><span class="p">][</span><span class="s1">&#39;bowtie&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39; -p &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_cpu</span><span class="p">)</span>
  <span class="n">cmd</span> <span class="o">+=</span> <span class="s1">&#39; -x &#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span>
  <span class="n">cmd</span> <span class="o">+=</span> <span class="s1">&#39; -1 &#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">i1</span> <span class="o">+</span> <span class="s1">&#39; -2 &#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">i2</span> <span class="o">+</span> <span class="s1">&#39; | &#39;</span>
  <span class="n">cmd</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;bin&#39;</span><span class="p">][</span><span class="s1">&#39;samtools&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39; view -bS - &#39;</span>
  <span class="n">cmd</span> <span class="o">+=</span> <span class="s1">&#39; | &#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;bin&#39;</span><span class="p">][</span><span class="s1">&#39;samtools&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39; view -u -f 12 -F 256 - | &#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;bin&#39;</span><span class="p">][</span><span class="s1">&#39;bedtools&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39; bamtofastq -i - -fq &#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">o1</span> <span class="o">+</span> <span class="s1">&#39; -fq2 &#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">o2</span>
  <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
  <span class="k">return</span> <span class="n">cmd</span>
</pre></div>
</div>
<div class="section" id="options">
<h4>Options<a class="headerlink" href="#options" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">i1</span></code>: R1 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">i2</span></code>: R2 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">db</span></code>: Bowtie database.</li>
<li><code class="docutils literal"><span class="pre">o1</span></code>: R1 output fastq file.</li>
<li><code class="docutils literal"><span class="pre">o2</span></code>: R2 output fastq file.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL]</li>
<li><code class="docutils literal"><span class="pre">n_cpu</span></code>: [INT] Number of CPU to use.</li>
<li><code class="docutils literal"><span class="pre">iter</span></code>: [SampleID]</li>
</ul>
</div>
</div>
<div class="section" id="demultiplex">
<h3>Demultiplex<a class="headerlink" href="#demultiplex" title="Permalink to this headline">¶</a></h3>
<a class="reference internal image-reference" href="_images/demultiplex_0.png"><img alt="_images/demultiplex_0.png" src="_images/demultiplex_0.png" style="width: 750.0px; height: 470.4px;" /></a>
<p>Warning: This demultiplex procedure is specifc to our sequencing methods.</p>
<p>The demultiplex.pl script uses <a class="reference external" href="https://github.com/marcelm/cutadapt/">cutadapt</a> to demultiplex and trim step by step sequences used by the sequencing technology and the dsRNA extraction protocol.</p>
<p>Our extraction and sequencing protocol:</p>
<p>Informations about indexes, common sequences, and sample ids are stored in the map.txt file which must be tab delimited file.
The script produce lots of temporary file (*_step_*) that can be deleted at the end of the execution.
Each step produce 3 type of files:</p>
<dl class="docutils">
<dt>.info:</dt>
<dd>the matching information.</dd>
<dt>.log:</dt>
<dd>the execution information.</dd>
<dt>.out:</dt>
<dd>the fastq file.</dd>
</dl>
<div class="section" id="id1">
<h4>Options<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">i1</span></code>: R1 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">i2</span></code>: R2 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">adapters</span></code>: Fasta file of adapters.</li>
<li><code class="docutils literal"><span class="pre">middle</span></code>: Check for MIDs in middle of the reads and 1 : trim the reads or 2: exclude the read.</li>
<li><code class="docutils literal"><span class="pre">min_qual</span></code>: Trim the read if the quality is below this threshold.</li>
<li><code class="docutils literal"><span class="pre">polyA</span></code>: Trim poly-A tail.</li>
<li><code class="docutils literal"><span class="pre">min_len</span></code>: Exclude reads under this size threshold.</li>
<li><code class="docutils literal"><span class="pre">iter</span></code>: [SampleID]</li>
</ul>
</div>
<div class="section" id="step-01-5-index-search">
<h4>Step 01: 5’ index search<a class="headerlink" href="#step-01-5-index-search" title="Permalink to this headline">¶</a></h4>
<a class="reference internal image-reference" href="_images/demultiplex_1.png"><img alt="_images/demultiplex_1.png" src="_images/demultiplex_1.png" style="width: 1000.0px; height: 95.6px;" /></a>
<div class="highlight-perl"><div class="highlight"><pre><span></span><span class="n">_launchCutAdapt</span><span class="p">(</span><span class="nv">$self</span><span class="p">,</span><span class="nv">$files</span><span class="o">-&gt;</span><span class="p">{</span><span class="mi">1</span><span class="p">},</span> <span class="nv">$self</span><span class="o">-&gt;</span><span class="p">{</span><span class="nb">index</span><span class="p">},</span> <span class="nv">$tmp_file_prefix</span> <span class="o">.</span> <span class="s">&quot;_step.01_R1&quot;</span><span class="p">,</span><span class="s">&#39;d&#39;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;-g&#39;</span><span class="p">,</span><span class="s">&#39;0&#39;</span><span class="p">,</span><span class="s">&#39;1&#39;</span><span class="p">,</span><span class="s">&#39;0.8&#39;</span><span class="p">);</span>
</pre></div>
</div>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">'d','k'</span></code>: This step discard untrimmed reads, so reads that do not contain indexes are excluded, and keep trimmed reads.</li>
<li><code class="docutils literal"><span class="pre">'-g'</span></code>: search indexes in 5’.</li>
<li><code class="docutils literal"><span class="pre">'0'</span></code>: no errors allowed in the index sequence.</li>
<li><code class="docutils literal"><span class="pre">'1'</span></code>: search for one index in each read.</li>
<li><code class="docutils literal"><span class="pre">'0.8'</span></code>: 80% of the index length to be considered as a match.</li>
</ul>
</div>
<div class="section" id="step-02-5-common-sequence">
<h4>Step 02: 5’ common sequence<a class="headerlink" href="#step-02-5-common-sequence" title="Permalink to this headline">¶</a></h4>
<a class="reference internal image-reference" href="_images/demultiplex_2.png"><img alt="_images/demultiplex_2.png" src="_images/demultiplex_2.png" style="width: 1000.0px; height: 110.6px;" /></a>
<div class="highlight-perl"><div class="highlight"><pre><span></span><span class="n">_launchCutAdapt</span><span class="p">(</span><span class="nv">$self</span><span class="p">,</span><span class="nv">$files</span><span class="o">-&gt;</span><span class="p">{</span><span class="mi">1</span><span class="p">},</span><span class="nv">$self</span><span class="o">-&gt;</span><span class="p">{</span><span class="n">_common</span><span class="p">},</span><span class="nv">$tmp_file_prefix</span> <span class="o">.</span> <span class="s">&quot;_step.02_R1&quot;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;-g&#39;</span><span class="p">,</span><span class="s">&#39;0.1&#39;</span><span class="p">,</span><span class="nb">scalar</span><span class="p">(</span><span class="nb">keys</span><span class="p">(</span><span class="nv">%</span><span class="p">{</span><span class="nv">$self</span><span class="o">-&gt;</span><span class="p">{</span><span class="n">_common</span><span class="p">}})),</span><span class="s">&#39;0.7&#39;</span><span class="p">);</span>
</pre></div>
</div>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">'k','k'</span></code>: Keep reads that contains or not the common part.</li>
<li><code class="docutils literal"><span class="pre">'-g'</span></code>: search in 5’ part.</li>
<li><code class="docutils literal"><span class="pre">'0.1'</span></code>: 10% of sequencing errors.</li>
<li><code class="docutils literal"><span class="pre">scalar(keys(%{$self-&gt;{_common}}))</span></code>: will search as many common part as provided.</li>
<li><code class="docutils literal"><span class="pre">'0.7'</span></code>: 70% of the common part length to be considered as a match.</li>
</ul>
</div>
<div class="section" id="step-03-5-common-sequence-fragments">
<h4>Step 03: 5’ common sequence fragments<a class="headerlink" href="#step-03-5-common-sequence-fragments" title="Permalink to this headline">¶</a></h4>
<a class="reference internal image-reference" href="_images/demultiplex_3.png"><img alt="_images/demultiplex_3.png" src="_images/demultiplex_3.png" style="width: 1000.0px; height: 109.4px;" /></a>
<div class="highlight-perl"><div class="highlight"><pre><span></span><span class="n">_launchCutAdapt</span><span class="p">(</span><span class="nv">$self</span><span class="p">,</span><span class="nv">$files</span><span class="o">-&gt;</span><span class="p">{</span><span class="mi">1</span><span class="p">},</span><span class="nv">$self</span><span class="o">-&gt;</span><span class="p">{</span><span class="n">_common</span><span class="p">},</span><span class="nv">$tmp_file_prefix</span> <span class="o">.</span> <span class="s">&quot;_step.021_R1&quot;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;-g&#39;</span><span class="p">,</span><span class="s">&#39;0.2&#39;</span><span class="p">,</span><span class="nb">scalar</span><span class="p">(</span><span class="nb">keys</span><span class="p">(</span><span class="nv">%</span><span class="p">{</span><span class="nv">$self</span><span class="o">-&gt;</span><span class="p">{</span><span class="n">_common</span><span class="p">}})),</span><span class="s">&#39;0.5&#39;</span><span class="p">);</span>
</pre></div>
</div>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">'k','k'</span></code>: Keep reads that contains or not the common part.</li>
<li><code class="docutils literal"><span class="pre">'-g'</span></code>: search in 5’ part.</li>
<li><code class="docutils literal"><span class="pre">'0.2'</span></code>: 20% of sequencing errors.</li>
<li><code class="docutils literal"><span class="pre">scalar(keys(%{$self-&gt;{_common}}))</span></code>: will search as many common part as provided.</li>
<li><code class="docutils literal"><span class="pre">'0.5'</span></code>: 50% of the common part length to be considered as a match.</li>
</ul>
</div>
<div class="section" id="step-04-trimming-sequencing-adapters">
<h4>Step 04: Trimming sequencing adapters<a class="headerlink" href="#step-04-trimming-sequencing-adapters" title="Permalink to this headline">¶</a></h4>
<a class="reference internal image-reference" href="_images/demultiplex_4.png"><img alt="_images/demultiplex_4.png" src="_images/demultiplex_4.png" style="width: 1000.0px; height: 118.2px;" /></a>
<div class="highlight-perl"><div class="highlight"><pre><span></span><span class="n">_launchCutAdapt</span><span class="p">(</span><span class="nv">$self</span><span class="p">,</span><span class="nv">$files</span><span class="o">-&gt;</span><span class="p">{</span><span class="mi">1</span><span class="p">},</span><span class="nv">$self</span><span class="o">-&gt;</span><span class="p">{</span><span class="n">illuminaAdapter</span><span class="p">},</span><span class="nv">$tmp_file_prefix</span> <span class="o">.</span> <span class="s">&quot;_step.03_R1&quot;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;-b&#39;</span><span class="p">,</span><span class="s">&#39;0.2&#39;</span><span class="p">,</span><span class="nb">scalar</span><span class="p">(</span><span class="nb">keys</span><span class="p">(</span><span class="nv">%</span><span class="p">{</span><span class="nv">$self</span><span class="o">-&gt;</span><span class="p">{</span><span class="n">illuminaAdapter</span><span class="p">}})),</span><span class="s">&#39;0.6&#39;</span><span class="p">);</span>
</pre></div>
</div>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">'k','k'</span></code>: Keep reads that contains or not the common part.</li>
<li><code class="docutils literal"><span class="pre">'-b'</span></code>: search adapters anywhere in the read.</li>
<li><code class="docutils literal"><span class="pre">'0.2'</span></code>: 20% of sequencing errors.</li>
<li><code class="docutils literal"><span class="pre">scalar(keys(%{$self-&gt;{illuminaAdapter}}))</span></code>: will search as many adapters as provided.</li>
<li><code class="docutils literal"><span class="pre">'0.6'</span></code>: 60% of the adapters length to be considered as a match.</li>
</ul>
</div>
<div class="section" id="step-05-search-for-hybrid-reads">
<h4>Step 05: Search for hybrid reads<a class="headerlink" href="#step-05-search-for-hybrid-reads" title="Permalink to this headline">¶</a></h4>
<a class="reference internal image-reference" href="_images/demultiplex_5.png"><img alt="_images/demultiplex_5.png" src="_images/demultiplex_5.png" style="width: 1000.0px; height: 126.0px;" /></a>
<p>This step is really specific to our extraction method since very short DNA fragment can be link together during the aspecific adapters ligation step of the Illumina kits. This creating reads composed of two different PCR product. Thus our program search for index sequence in the middle of the read and trim it to keep the 5’ part or exlude the read.
The research is done both on provided indexes sequences and reverse complement of thoses sequences.</p>
<p><code class="docutils literal"><span class="pre">-middle</span> <span class="pre">[1|2]</span></code>   Search for common tag in the middle of the read. 1: trim the read. 2: exclude the read.</p>
<div class="highlight-perl"><div class="highlight"><pre><span></span><span class="n">_launchCutAdapt</span><span class="p">(</span><span class="nv">$self</span><span class="p">,</span><span class="nv">$files</span><span class="o">-&gt;</span><span class="p">{</span><span class="mi">1</span><span class="p">},</span><span class="nv">$h</span><span class="p">,</span><span class="nv">$tmp_file_prefix</span> <span class="o">.</span> <span class="s">&quot;_step.04_R1&quot;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;-b&#39;</span><span class="p">,</span><span class="s">&#39;0.1&#39;</span><span class="p">,</span><span class="s">&#39;1&#39;</span><span class="p">,</span><span class="s">&#39;0.5&#39;</span><span class="p">);</span>
</pre></div>
</div>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">'k','k'</span></code>: Keep reads that contains or not the index , or <code class="docutils literal"><span class="pre">'k','d'</span></code> if the <code class="docutils literal"><span class="pre">-middle</span></code> option is provided.</li>
<li><code class="docutils literal"><span class="pre">'-b'</span></code>: search adapters anywhere in the read.</li>
<li><code class="docutils literal"><span class="pre">'0.1'</span></code>: 20% of sequencing errors.</li>
<li><code class="docutils literal"><span class="pre">'1'</span></code>: search for one index in each read.</li>
<li><code class="docutils literal"><span class="pre">'0.5'</span></code>: 50% of the adapters length to be considered as a match.</li>
</ul>
</div>
<div class="section" id="step-06-search-for-polya-optional">
<h4>Step 06: Search for polyA (optional)<a class="headerlink" href="#step-06-search-for-polya-optional" title="Permalink to this headline">¶</a></h4>
<a class="reference internal image-reference" href="_images/demultiplex_6.png"><img alt="_images/demultiplex_6.png" src="_images/demultiplex_6.png" style="width: 750.0px; height: 150.0px;" /></a>
<p>In Illumina technology, if the sequencing matrix is too short compared to the sequencing length, the sequencing machine adds a bunch of A’s and then radom sequence.</p>
<div class="highlight-perl"><div class="highlight"><pre><span></span><span class="n">_launchCutAdapt</span><span class="p">(</span><span class="nv">$self</span><span class="p">,</span><span class="nv">$files</span><span class="o">-&gt;</span><span class="p">{</span><span class="mi">1</span><span class="p">},</span> <span class="nv">$h</span><span class="p">,</span> <span class="nv">$tmp_file_prefix</span> <span class="o">.</span> <span class="s">&quot;_step.05_R1&quot;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;-a&#39;</span><span class="p">,</span><span class="s">&#39;0&#39;</span><span class="p">,</span><span class="s">&#39;1&#39;</span><span class="p">,</span><span class="s">&#39;0.8&#39;</span><span class="p">);</span>
</pre></div>
</div>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">'k','k'</span></code>: Keep reads that contains or not the index , or <code class="docutils literal"><span class="pre">'k','d'</span></code> if the <code class="docutils literal"><span class="pre">-middle</span></code> option is provided.</li>
<li><code class="docutils literal"><span class="pre">'-a'</span></code>: search in the 3’ end.</li>
<li><code class="docutils literal"><span class="pre">'0'</span></code>: no sequencing errors.</li>
<li><code class="docutils literal"><span class="pre">'1'</span></code>: search for one index in each read.</li>
<li><code class="docutils literal"><span class="pre">'0.8'</span></code>: 80% of the polyA length to be considered as a match.</li>
</ul>
</div>
</div>
<div class="section" id="assembly">
<h3>Assembly<a class="headerlink" href="#assembly" title="Permalink to this headline">¶</a></h3>
<p>This module can launch two assemblers, <a class="reference external" href="https://github.com/loneknightpy/idba/">IDBA</a> and <a class="reference external" href="http://bioinf.spbau.ru/en/spades3.7/">MetaSpades</a>.</p>
<p>Foreach assembler, the module convert reads files to the proper format, launch the assembly in a separate directory, rename scaffolds identifier and move results file to the sample root directory.</p>
<div class="section" id="id2">
<h4>Options<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h4>
</div>
</div>
<div class="section" id="map">
<h3>Map<a class="headerlink" href="#map" title="Permalink to this headline">¶</a></h3>
<p>This module uses bowtie2, samtools and readPerContig.pl script to map reads back on the assembly and count for each scaffold the number of reads aligned resulting a simple two column file scaffoldID and nb_reads used by other modules.</p>
<div class="section" id="id3">
<h4>Options<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">contigs</span></code>: fasta file of contigs to map reads on. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">i1</span></code>: R1 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">i2</span></code>: R2 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">ising</span></code>: singletons fastq file</li>
<li><code class="docutils literal"><span class="pre">n_cpu</span></code>: [INT] number of CPU to use.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL] use SGE scheduler.</li>
<li><code class="docutils literal"><span class="pre">bam</span></code>: BAM file name.</li>
<li><code class="docutils literal"><span class="pre">rn</span></code>: output file name.</li>
</ul>
</div>
</div>
<div class="section" id="normalization">
<h3>Normalization<a class="headerlink" href="#normalization" title="Permalink to this headline">¶</a></h3>
<p>This module randomly select NUM reads from paired-files.</p>
<div class="section" id="id4">
<h4>Options<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">i1</span></code>: R1 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">i2</span></code>: R2 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">o1</span></code>: Output R1 normalized file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">o2</span></code>: Output R2 normalized file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">num</span></code>: [INT] Number of reads to randomly select. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">iter</span></code>: Iteration on [sample, library].</li>
<li><code class="docutils literal"><span class="pre">n_cpu</span></code>: [INT] number of CPU to use.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL] use SGE scheduler.</li>
</ul>
</div>
</div>
<div class="section" id="diamond">
<h3>Diamond<a class="headerlink" href="#diamond" title="Permalink to this headline">¶</a></h3>
<p>This module launches Diamond similarity search on reads and produce an XML file simalar to what Blast does so it can be treated by the Blast2ecsv module and so on.</p>
<div class="section" id="id5">
<h4>Options<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">i1</span></code>: R1 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">i2</span></code>: R2 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">db</span></code>: Values are defined in the parameters.yaml file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">ising</span></code>: singletons fastq file</li>
<li><code class="docutils literal"><span class="pre">n_cpu</span></code>: [INT] number of CPU to use.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL] use SGE scheduler.</li>
<li><code class="docutils literal"><span class="pre">sensitive</span></code>: [BOOL]</li>
<li><code class="docutils literal"><span class="pre">more_sensitive</span></code>: [BOOL]</li>
<li><code class="docutils literal"><span class="pre">out</span></code>: XML output file</li>
<li><code class="docutils literal"><span class="pre">score</span></code>: Report matches above this score.</li>
<li><code class="docutils literal"><span class="pre">max_target_seqs</span></code>: Maximum match per query sequences.</li>
<li><code class="docutils literal"><span class="pre">evalue</span></code>: Min e-value.</li>
<li><code class="docutils literal"><span class="pre">identity</span></code>: Report matches above this identity percent. 0 &gt; X &gt; 100.</li>
<li><code class="docutils literal"><span class="pre">qov</span></code>: Query overlap.</li>
<li><code class="docutils literal"><span class="pre">hov</span></code>: Hit overlap.</li>
</ul>
</div>
</div>
<div class="section" id="diamond2blast">
<h3>Diamond2Blast<a class="headerlink" href="#diamond2blast" title="Permalink to this headline">¶</a></h3>
<div class="section" id="id6">
<h4>Options<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">i</span></code>: CSV file with DIAMOND results. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">contigs</span></code>: Fasta file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">out</span></code>: XML output file.</li>
<li><code class="docutils literal"><span class="pre">type</span></code>: Blast type. [‘tblastx’,’blastx’,’blastn’,’blastp’,’rpstblastn’]. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">db</span></code>: Values are defined in the parameters.yaml file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">evalue</span></code>: Min e-value.</li>
<li><code class="docutils literal"><span class="pre">server</span></code>: [‘enki’,’genologin’,’avakas’] Values are defined in the parameters.yaml file.</li>
<li><code class="docutils literal"><span class="pre">n_cpu</span></code>: [INT] number of CPU to use.</li>
<li><code class="docutils literal"><span class="pre">tc</span></code>: Number of task launched at the same time on SGE.</li>
<li><code class="docutils literal"><span class="pre">num_chunk</span></code>: Number of chunks to split the original fasta file for parallel execution.</li>
<li><code class="docutils literal"><span class="pre">max_target_seqs</span></code>: Maximum match per query sequences.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL] use SGE scheduler.</li>
</ul>
</div>
</div>
<div class="section" id="blast">
<h3>Blast<a class="headerlink" href="#blast" title="Permalink to this headline">¶</a></h3>
<p>This module launches all type of Blast on local machine or distant servers. This module has been developped for our own local machines and servers, but it can be easly modified to fit your needs.</p>
<p>This module mainly depends on the parameters.yaml file and the blast_launch.py script which has to be present on the server you want to use and modified to fit your server configuration.</p>
<div class="section" id="id7">
<h4>Options<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">contigs</span></code>: Fasta file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">db</span></code>: Values are defined in the parameters.yaml file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">type</span></code>: Blast type. [‘tblastx’,’blastx’,’blastn’,’blastp’,’rpstblastn’]. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">n_cpu</span></code>: [INT] number of CPU to use.</li>
<li><code class="docutils literal"><span class="pre">tc</span></code>: Number of task launched at the same time on SGE. (Experimental, works on Genotoul)</li>
<li><code class="docutils literal"><span class="pre">max_target_seqs</span></code>: Maximum match per query sequences.</li>
<li><code class="docutils literal"><span class="pre">num_chunk</span></code>: Number of chunks to split the original fasta file for parallel execution.</li>
<li><code class="docutils literal"><span class="pre">out</span></code>: Output file name.</li>
<li><code class="docutils literal"><span class="pre">server</span></code>: [‘enki’,’genologin’,’avakas’] Values are defined in the parameters.yaml file.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL] use SGE scheduler.</li>
</ul>
<p>This module is able to launch Blast instance on distant servers if the database and the blast_launch.py script is present on the server. Then you have to edit the parameters.yaml file to fit your configuration. The script has been developped to use two computer cluster, Avakas (PBS + Torque) and Genotoul (SGE) but each cluster has its own configuration so you may have to modify this script to adapt it to your configuration.</p>
</div>
</div>
<div class="section" id="blast2ecsv">
<h3>Blast2ecsv<a class="headerlink" href="#blast2ecsv" title="Permalink to this headline">¶</a></h3>
<p>This module parse Blast xml outputs, filter matches on different criteria and link Accession number to NCBI taxonomy.</p>
<div class="section" id="id8">
<h4>Options<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">b</span></code>: Blast file.</li>
<li><code class="docutils literal"><span class="pre">if</span></code>: Input format [‘xml’,’m8’]</li>
<li><code class="docutils literal"><span class="pre">out</span></code>: Output file name.</li>
<li><code class="docutils literal"><span class="pre">evalue</span></code>: Min e-value.</li>
<li><code class="docutils literal"><span class="pre">fhit</span></code>: Only report first hit.</li>
<li><code class="docutils literal"><span class="pre">fhsp</span></code>: Only report first hsp.</li>
<li><code class="docutils literal"><span class="pre">pm</span></code>:</li>
<li><code class="docutils literal"><span class="pre">r</span></code>: Reduced taxonomy. Report only 5 consistent rank.</li>
<li><code class="docutils literal"><span class="pre">vs</span></code>: Only report sequences when match is virus or viroids.</li>
<li><code class="docutils literal"><span class="pre">rn</span></code>: Read number. File created by the Map module.</li>
<li><code class="docutils literal"><span class="pre">type</span></code>: Blast type. [‘tblastx’,’blastx’,’blastn’,’blastp’,’rpstblastn’]</li>
<li><code class="docutils literal"><span class="pre">score</span></code>: Report matches above this score.</li>
<li><code class="docutils literal"><span class="pre">identity</span></code>: Report matches above this identity percent. 0 &gt; X &gt; 100.</li>
<li><code class="docutils literal"><span class="pre">qov</span></code>: Query overlap.</li>
<li><code class="docutils literal"><span class="pre">hov</span></code>: Hit overlap.</li>
<li><code class="docutils literal"><span class="pre">pd</span></code>: Parse description. Useful when the query ID is stored in the dscription field in the XML file.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL] use SGE scheduler.</li>
</ul>
</div>
</div>
<div class="section" id="ecsv2excel">
<h3>Ecsv2excel<a class="headerlink" href="#ecsv2excel" title="Permalink to this headline">¶</a></h3>
<p>This module aggregates multiple ecsv file to create a colored XLSX file.
It launches the ecsv2krona.pl script.</p>
<div class="section" id="id9">
<h4>Options<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">b[INT]</span></code>: CSV Blast file from 1 to 10.</li>
<li><code class="docutils literal"><span class="pre">out</span></code>: Outpuy file name.</li>
<li><code class="docutils literal"><span class="pre">r</span></code>: RPSBLAST csv file.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL] use SGE scheduler.</li>
</ul>
</div>
</div>
<div class="section" id="ecsv2krona">
<h3>Ecsv2krona<a class="headerlink" href="#ecsv2krona" title="Permalink to this headline">¶</a></h3>
<p>This module launch the ecsv2krona.pl script. It will aggregate multiple ecsv file into one Krona html file.</p>
<div class="section" id="id10">
<h4>Options<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">b</span></code>: [INT] CSV Blast file.</li>
<li><code class="docutils literal"><span class="pre">id</span></code>: [INT] ID wanted corresponding to the Blast file.</li>
<li><code class="docutils literal"><span class="pre">x</span></code>: [INT] XML Blast file. If used, this file will be split by species and link in the Krona file.</li>
<li><code class="docutils literal"><span class="pre">out</span></code>: Output file name.</li>
<li><code class="docutils literal"><span class="pre">data</span></code>: [‘both’,’reads’,’contigs’,’none’]</li>
<li><code class="docutils literal"><span class="pre">r</span></code>: Use reduced taxonomy.</li>
<li><code class="docutils literal"><span class="pre">c</span></code>: [‘identity’,’taxid’,’none’]</li>
<li><code class="docutils literal"><span class="pre">iter</span></code>: [‘global’]</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL] use SGE scheduler.</li>
</ul>
</div>
</div>
<div class="section" id="rps2ecsv">
<h3>Rps2ecsv<a class="headerlink" href="#rps2ecsv" title="Permalink to this headline">¶</a></h3>
<p>This module launch rps2ecsv.pl script for each sample.
This module parse XML files from rpsblast and create csv file as a result.</p>
<div class="section" id="id11">
<h4>Options<a class="headerlink" href="#id11" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">b</span></code>: RPSBLAST XML file.</li>
<li><code class="docutils literal"><span class="pre">contigs</span></code>: Fasta file.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL] use SGE scheduler.</li>
<li><code class="docutils literal"><span class="pre">out</span></code>: Output file name.</li>
<li><code class="docutils literal"><span class="pre">evalue</span></code>: e-value threshold.</li>
</ul>
</div>
</div>
<div class="section" id="rps2tree">
<h3>Rps2tree<a class="headerlink" href="#rps2tree" title="Permalink to this headline">¶</a></h3>
<p>This module launch rps2tree.pl script for all sample provided.
This module generates Operational Taxonomic Unit (OTU) based on RPS-Blast results.
For each CDD motifs, contigs are clustered together based on matrix distance.</p>
<div class="section" id="id12">
<h4>Options<a class="headerlink" href="#id12" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">pfam</span></code>: CSV file from Rps2ecsv.</li>
<li><code class="docutils literal"><span class="pre">contigs</span></code>: Fasta file.</li>
<li><code class="docutils literal"><span class="pre">ecsv</span></code>: CSV file from Blast2ecsv.</li>
<li><code class="docutils literal"><span class="pre">out</span></code>: Output file name.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL]</li>
<li><code class="docutils literal"><span class="pre">viral_portion</span></code>: Minimum percentage of viral sequence in a domain to be selected.</li>
<li><code class="docutils literal"><span class="pre">min_prot</span></code>: Minimum protein length to be included in a tree.</li>
<li><code class="docutils literal"><span class="pre">perc</span></code>: Percentage of identity. Threshold set to define OTU.</li>
<li><code class="docutils literal"><span class="pre">iter</span></code>: [‘global’]</li>
</ul>
</div>
</div>
<div class="section" id="automapper">
<h3>AutoMapper<a class="headerlink" href="#automapper" title="Permalink to this headline">¶</a></h3>
<p>This module launch autoMapper.pl script for every sample.</p>
<div class="section" id="id13">
<h4>Options<a class="headerlink" href="#id13" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">contigs</span></code>: Fasta file.</li>
<li><code class="docutils literal"><span class="pre">ecsv</span></code>: CSV file from Blast2ecsv.</li>
<li><code class="docutils literal"><span class="pre">i1</span></code>: R1 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">i2</span></code>: R2 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">out</span></code>: Output folder.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL]</li>
<li><code class="docutils literal"><span class="pre">ref</span></code>: Blast database name.</li>
</ul>
</div>
</div>
<div class="section" id="blast2hits">
<h3>Blast2hits<a class="headerlink" href="#blast2hits" title="Permalink to this headline">¶</a></h3>
<p>This takes multiple CSV Blast file from Blast2ecsv and draw histograms for by taxonomy.</p>
<div class="section" id="id14">
<h4>Options<a class="headerlink" href="#id14" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">b[INT]</span></code> CSV Blast file from Blast2ecsv</li>
<li><code class="docutils literal"><span class="pre">id[INT]</span></code> ID associated with the Blast file.</li>
<li><code class="docutils literal"><span class="pre">iter</span></code> [global]</li>
<li><code class="docutils literal"><span class="pre">sge</span></code> [BOOL]</li>
<li><code class="docutils literal"><span class="pre">out</span></code> Output file name.</li>
</ul>
</div>
</div>
</div>
<span id="document-execution"></span><div class="section" id="example-execution">
<h2>Example execution<a class="headerlink" href="#example-execution" title="Permalink to this headline">¶</a></h2>
<p>Create a directory for your experiment:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="n">test_virAnnot</span>
<span class="n">cd</span> <span class="n">test_virAnnot</span>
</pre></div>
</div>
<p>Copy example read files, Illumina adapters fasta file, the sample id mapping file, the step and parameter file:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">cp</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">virAnnot</span><span class="o">/</span><span class="n">examples</span><span class="o">/</span><span class="n">reads</span><span class="o">/*.</span><span class="n">fq</span> <span class="o">.</span>
<span class="n">cp</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">virAnnot</span><span class="o">/</span><span class="n">examples</span><span class="o">/</span><span class="n">adapters</span><span class="o">.</span><span class="n">fa</span> <span class="o">.</span>
<span class="n">cp</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">virAnnot</span><span class="o">/</span><span class="n">examples</span><span class="o">/</span><span class="nb">map</span><span class="o">.</span><span class="n">txt</span> <span class="o">.</span>
<span class="n">cp</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">virAnnot</span><span class="o">/</span><span class="n">examples</span><span class="o">/</span><span class="n">step</span><span class="o">.</span><span class="n">yaml</span> <span class="o">.</span>
<span class="n">cp</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">virAnnot</span><span class="o">/</span><span class="n">examples</span><span class="o">/</span><span class="n">parameters</span><span class="o">.</span><span class="n">yaml</span> <span class="o">.</span>
</pre></div>
</div>
<p>You have to modify this file to fit your configuration.</p>
<p>This example contains all modules and options available and must be used as a template for your own analysis.</p>
<div class="section" id="step-readsoustraction">
<h3>Step <strong>ReadSoustraction</strong><a class="headerlink" href="#step-readsoustraction" title="Permalink to this headline">¶</a></h3>
<p>This module uses bowtie2 to map reads against nucleotide sequence and Samtools to extract unmapped pairs.</p>
<p>Corresponding <code class="docutils literal"><span class="pre">step.yaml</span></code> section:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">ReadSoustraction_phiX</span><span class="p">:</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">file1</span><span class="p">)</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">file2</span><span class="p">)</span>
  <span class="n">db</span><span class="p">:</span> <span class="n">phiX</span>
  <span class="n">o1</span><span class="p">:</span> <span class="p">(</span><span class="n">library</span><span class="p">)</span><span class="n">_phiX</span><span class="o">.</span><span class="n">r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">o2</span><span class="p">:</span> <span class="p">(</span><span class="n">library</span><span class="p">)</span><span class="n">_phiX</span><span class="o">.</span><span class="n">r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">5</span>
  <span class="nb">iter</span><span class="p">:</span> <span class="n">library</span>
</pre></div>
</div>
<div class="highlight-bash"><div class="highlight"><pre><span></span>virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n ReadSoustraction_phiX
</pre></div>
</div>
</div>
<div class="section" id="step-demultiplex">
<h3>Step <strong>Demultiplex</strong><a class="headerlink" href="#step-demultiplex" title="Permalink to this headline">¶</a></h3>
<p>This module uses cutadapt demultiplex reads from library and also trim reads from adapters and chimeric reads.
Each demultiplexing step are described in the module section.
Corresponding <code class="docutils literal"><span class="pre">step.yaml</span></code> section:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Demultiplex</span><span class="p">:</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">library</span><span class="p">)</span><span class="n">_phiX</span><span class="o">.</span><span class="n">r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">library</span><span class="p">)</span><span class="n">_phiX</span><span class="o">.</span><span class="n">r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">adapters</span><span class="p">:</span> <span class="n">adapters</span><span class="o">.</span><span class="n">fna</span>
  <span class="n">middle</span><span class="p">:</span> <span class="mi">1</span>
  <span class="n">min_qual</span><span class="p">:</span> <span class="mi">20</span>
  <span class="n">polyA</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">min_len</span><span class="p">:</span> <span class="mi">70</span>
  <span class="nb">iter</span><span class="p">:</span> <span class="n">library</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
</pre></div>
</div>
<div class="highlight-bash"><div class="highlight"><pre><span></span>virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Demultiplex
</pre></div>
</div>
</div>
<div class="section" id="step-demultiplexhtml">
<h3>Step <strong>DemultiplexHtml</strong><a class="headerlink" href="#step-demultiplexhtml" title="Permalink to this headline">¶</a></h3>
<p>This module gather all <code class="docutils literal"><span class="pre">*_demultiplex.stats.csv</span></code> files and create and html report.
Corresponding <code class="docutils literal"><span class="pre">step.yaml</span></code> section:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">DemultiplexHtml</span><span class="p">:</span>
  <span class="n">csv</span><span class="p">:</span> <span class="p">(</span><span class="n">library</span><span class="p">)</span><span class="n">_demultiplex</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">csv</span>
  <span class="nb">id</span><span class="p">:</span> <span class="p">(</span><span class="n">library</span><span class="p">)</span>
  <span class="n">out</span><span class="p">:</span> <span class="n">stat_demultiplex</span>
  <span class="nb">iter</span><span class="p">:</span> <span class="k">global</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
</pre></div>
</div>
<div class="highlight-bash"><div class="highlight"><pre><span></span>virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n DemultiplexHtml
</pre></div>
</div>
<p>Output example:</p>
<iframe src="_static/stat_demultiplex/index.html" height="400px" width="100%"></iframe></div>
<div class="section" id="step-diamond">
<h3>Step <strong>Diamond</strong><a class="headerlink" href="#step-diamond" title="Permalink to this headline">¶</a></h3>
<p>This module launch Diamond similarity search for reads and produce an XML file per sample.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">16</span>
<span class="n">Map_spades</span><span class="p">:</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_spades</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">bam</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_spades</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">bam</span>
  <span class="n">rn</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_spades</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">rn</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">16</span>
<span class="n">Diamond</span><span class="p">:</span>
</pre></div>
</div>
<div class="highlight-bash"><div class="highlight"><pre><span></span>virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Diamond
</pre></div>
</div>
</div>
<div class="section" id="step-assembly">
<h3>Step <strong>Assembly</strong><a class="headerlink" href="#step-assembly" title="Permalink to this headline">¶</a></h3>
<p>This module simply launch udba_ud and metaspades assemblers in each sample folder, rename scaffolds id and move the resulting fasta file.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">5</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">drVM</span><span class="p">:</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">20</span>
  <span class="n">identity</span><span class="p">:</span> <span class="mi">70</span>
  <span class="n">min_len</span><span class="p">:</span> <span class="mi">300</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Assembly_idba</span><span class="p">:</span>
  <span class="n">prog</span><span class="p">:</span> <span class="n">idba</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">5</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r2</span><span class="o">.</span><span class="n">fq</span>
</pre></div>
</div>
<p>Test both idba and spades:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Assembly_idba
virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Assembly_spades
</pre></div>
</div>
<p>Example of idba assembly:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">ds2015</span><span class="o">-</span><span class="mi">149</span><span class="n">_0</span>
<span class="n">GTGTAAGGTGGTGAAGG</span><span class="o">...</span>
<span class="o">&gt;</span><span class="n">ds2015</span><span class="o">-</span><span class="mi">149</span><span class="n">_1</span>
<span class="n">CCTGCGAATTGGGCCAA</span><span class="o">...</span>
</pre></div>
</div>
</div>
<div class="section" id="step-map">
<h3>Step <strong>Map</strong><a class="headerlink" href="#step-map" title="Permalink to this headline">¶</a></h3>
<p>This module uses bowtie2 to map reads back on assemblies and samtools will count reads per contig.</p>
<p>Step file:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Assembly_spades</span><span class="p">:</span>
  <span class="n">prog</span><span class="p">:</span> <span class="n">spades</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">5</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_spades</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Map_idba</span><span class="p">:</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">i1</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r1</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">i2</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_truePairs_r2</span><span class="o">.</span><span class="n">fq</span>
  <span class="n">bam</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">bam</span>
  <span class="n">rn</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">rn</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
</pre></div>
</div>
<p>Command:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Map_idba
virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Map_spades
</pre></div>
</div>
<p>Output a two column tabular file, column 1 sequence ID, column 2 number of reads.
Example of <code class="docutils literal"><span class="pre">.rn</span></code> file produce:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">ds2015</span><span class="o">-</span><span class="mi">149</span><span class="n">_0</span>    <span class="mi">1179</span>
<span class="n">ds2015</span><span class="o">-</span><span class="mi">149</span><span class="n">_1</span>    <span class="mi">444</span>
<span class="n">ds2015</span><span class="o">-</span><span class="mi">149</span><span class="n">_10</span>   <span class="mi">26</span>
<span class="n">ds2015</span><span class="o">-</span><span class="mi">149</span><span class="n">_11</span>   <span class="mi">44</span>
<span class="n">ds2015</span><span class="o">-</span><span class="mi">149</span><span class="n">_12</span>   <span class="mi">14</span>
<span class="n">ds2015</span><span class="o">-</span><span class="mi">149</span><span class="n">_13</span>   <span class="mi">4</span>
<span class="n">ds2015</span><span class="o">-</span><span class="mi">149</span><span class="n">_14</span>   <span class="mi">6</span>
</pre></div>
</div>
</div>
<div class="section" id="step-blast">
<h3>Step <strong>Blast</strong><a class="headerlink" href="#step-blast" title="Permalink to this headline">¶</a></h3>
<p>This module is able to launch Blast(s) against provided databases localy or remotely.
The script blast_launch.py must be present on distant servers and <code class="docutils literal"><span class="pre">parameter.yaml</span></code> modified to fit your servers.</p>
<p>Step file:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>  <span class="n">i</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">dmdx</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">dmdx2bltx</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">dmdx2bltx</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">xml</span>
  <span class="nb">type</span><span class="p">:</span> <span class="n">blastx</span>
  <span class="n">db</span><span class="p">:</span> <span class="n">nr</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.0001</span>
  <span class="n">server</span><span class="p">:</span> <span class="n">genologin</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">8</span>
  <span class="n">tc</span><span class="p">:</span> <span class="mi">50</span>
  <span class="n">num_chunk</span><span class="p">:</span> <span class="mi">1000</span>
  <span class="n">max_target_seqs</span><span class="p">:</span> <span class="mi">1</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Blast_allvirTX</span><span class="p">:</span>
  <span class="nb">type</span><span class="p">:</span> <span class="n">tblastx</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">db</span><span class="p">:</span> <span class="n">all_vir_nucl</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">all_vir</span><span class="o">.</span><span class="n">xml</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.0001</span>
  <span class="n">server</span><span class="p">:</span> <span class="n">genotoul</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">8</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">num_chunk</span><span class="p">:</span> <span class="mi">1000</span>
  <span class="n">tc</span><span class="p">:</span> <span class="mi">50</span>
<span class="n">Blast_nr</span><span class="p">:</span>
  <span class="nb">type</span><span class="p">:</span> <span class="n">blastx</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">db</span><span class="p">:</span> <span class="n">nr</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">bltx</span><span class="o">.</span><span class="n">nr</span><span class="o">.</span><span class="n">xml</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.0001</span>
  <span class="n">server</span><span class="p">:</span> <span class="n">genotoul</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">8</span>
  <span class="n">tc</span><span class="p">:</span> <span class="mi">50</span>
  <span class="n">num_chunk</span><span class="p">:</span> <span class="mi">1000</span>
  <span class="n">max_target_seqs</span><span class="p">:</span> <span class="mi">1</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
<span class="n">Blast_refvirTX</span><span class="p">:</span>
  <span class="nb">type</span><span class="p">:</span> <span class="n">tblastx</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">db</span><span class="p">:</span> <span class="n">refseq_vir_nucl</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">refseq_vir</span><span class="o">.</span><span class="n">xml</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.0001</span>
  <span class="n">server</span><span class="p">:</span> <span class="n">genotoul</span>
  <span class="n">n_cpu</span><span class="p">:</span> <span class="mi">8</span>
</pre></div>
</div>
<p>Commands:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Blast_nr
virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Blast_refvirTX
virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Blast_allvirTX
virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Blast_RPS
</pre></div>
</div>
</div>
<div class="section" id="step-blast2ecsv">
<h3>Step <strong>Blast2ecsv</strong><a class="headerlink" href="#step-blast2ecsv" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div>This module uses the XML file generated by the corresponding Blast module and the taxonomy contained in the SQLITE database to create a csv file containing match options, taxonomy string and sequences.</div></blockquote>
<div class="highlight-bash"><div class="highlight"><pre><span></span>virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Blast2ecsv_nr
virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Blast2ecsv_refvirTX
virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Blast2ecsv_allvirTX
</pre></div>
</div>
<p>Example output of ds2015-149_idba.scaffold.tbltx.all_vir.csv:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="c1">#algo   query_id        nb_reads        query_length    accession       description     organism        percentIdentity nb_hsps queryOverlap    hitOverlap      evalue  score   tax_id  taxonomy        sequence</span>
<span class="s2">&quot;TBLASTX&quot;</span>       <span class="s2">&quot;ds2015-149_52&quot;</span> <span class="s2">&quot;6&quot;</span>     <span class="s2">&quot;117&quot;</span>   <span class="s2">&quot;KX274275.1&quot;</span>    <span class="s2">&quot;Grapevine rupestris stem pitting associated virus isolate SK704 B, complete genome&quot;</span>    <span class="s2">&quot;Grapevine rupestris stem pitting-associated virus&quot;</span>     <span class="s2">&quot;95.8&quot;</span>  <span class="s2">&quot;3&quot;</span>     <span class="s2">&quot;100&quot;</span>   <span class="s2">&quot;3&quot;</span>     <span class="s2">&quot;7.55823333338424e-05&quot;</span>  <span class="s2">&quot;222.2257&quot;</span>      <span class="s2">&quot;196400&quot;</span>        <span class="s2">&quot;Viruses;ssRNA viruses;Betaflexiviridae;Foveavirus;Grapevine rupestris stem pitting-associated virus&quot;</span>   <span class="s2">&quot;GAACACTATGAACGACAACTGGAAATCTGAGCACGCTATAAACACCCACAAACTCAAACGTAGACAAAGCTTTAACTAAGTTATTCATAATAATCACACCATGCCAAACACTTAAGG&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="step-rps2ecsv">
<h3>Step <strong>Rps2ecsv</strong><a class="headerlink" href="#step-rps2ecsv" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div>This module uses the rpstblastn XML file and the PFAM taxonomy to annotate query sequences and produce a readable CSV file.</div></blockquote>
<div class="highlight-bash"><div class="highlight"><pre><span></span>virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Rps2ecsv
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span>  <span class="n">vs</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">out</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">tbltx</span><span class="o">.</span><span class="n">refseq_vir</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">sge</span><span class="p">:</span> <span class="kc">True</span>
  <span class="nb">type</span><span class="p">:</span> <span class="n">TBLASTX</span>
  <span class="n">score</span><span class="p">:</span> <span class="mi">50</span>
</pre></div>
</div>
<p>Example output of ds2015-149_idba.scaffold.rps.pfam.csv:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="c1">#query_id       query_length    cdd_id  hit_id  evalue  startQ  endQ    frame   description     superkingdom    no rank family  genus</span>
<span class="s2">&quot;ds2015-149_0&quot;</span>  <span class="s2">&quot;1428&quot;</span>  <span class="s2">&quot;pfam01443&quot;</span>     <span class="s2">&quot;gnl|CDD|279750&quot;</span>        <span class="s2">&quot;1.33194e-06&quot;</span>   <span class="s2">&quot;29&quot;</span>    <span class="s2">&quot;223&quot;</span>   <span class="s2">&quot;2&quot;</span>     <span class="s2">&quot;pfam01443, Viral_helicase1, Viral (Superfamily 1) RNA helicase.  Helicase activity for this family has been demonstrated and NTPase activity. This helicase has multiple roles at different stages of viral RNA replication, as dissected by mutational analysis.&quot;</span>     <span class="s2">&quot;Viruses(1.00);&quot;</span>        <span class="s2">&quot;ssRNA viruses(0.99);unclassified viruses(0.01);&quot;</span>       <span class="s2">&quot;Alphaflexiviridae(0.36);Virgaviridae(0.24);Betaflexiviridae(0.15);Tymoviridae(0.10);Bromoviridae(0.07);&quot;</span>       <span class="s2">&quot;Potexvirus(0.26);Allexivirus(0.10);Carlavirus(0.08);Tymovirus(0.08);Tobamovirus(0.08);&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="step-ecsv2excel">
<h3>Step <strong>Ecsv2excel</strong><a class="headerlink" href="#step-ecsv2excel" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div>This module takes all csv files and create a compile them in a single Excel file.</div></blockquote>
<div class="highlight-bash"><div class="highlight"><pre><span></span>virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Ecsv2excel
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span>  <span class="n">qov</span><span class="p">:</span> <span class="mi">50</span>
  <span class="n">hov</span><span class="p">:</span> <span class="mi">5</span>
<span class="n">Blast2ecsv_nr</span><span class="p">:</span>
  <span class="n">contigs</span><span class="p">:</span> <span class="p">(</span><span class="n">SampleID</span><span class="p">)</span><span class="n">_idba</span><span class="o">.</span><span class="n">scaffold</span><span class="o">.</span><span class="n">fa</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="mf">0.001</span>
  <span class="n">fhit</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">pm</span><span class="p">:</span> <span class="k">global</span>
</pre></div>
</div>
<p>Example output of ds2015-149_idba.scaffold.xlsx:</p>
<img alt="_images/ecsv2excel.png" src="_images/ecsv2excel.png" />
</div>
<div class="section" id="step-ecsv2krona">
<h3>Step <strong>Ecsv2krona</strong><a class="headerlink" href="#step-ecsv2krona" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div>This module uses CSV files from Blast2ecsv module to create Krona html file.</div></blockquote>
<div class="highlight-bash"><div class="highlight"><pre><span></span>virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Ecsv2krona
virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Ecsv2krona_dmd
</pre></div>
</div>
<p>Example output of Krona:</p>
<iframe src="_static/krona_blast/blast.global.krona.html" height="400px" width="100%"></iframe>
<iframe src="_static/krona_diamond/global_krona_dmd.html" height="400px" width="100%"></iframe></div>
<div class="section" id="step-automapper">
<h3>Step <strong>Automapper</strong><a class="headerlink" href="#step-automapper" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div>This module uses Blast CSV annotation file to select reference sequences, map query sequences and produce png of identity plot and alignment file in fasta format.</div></blockquote>
<div class="highlight-bash"><div class="highlight"><pre><span></span>virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Automapper_nr
virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Automapper_allvirTX
virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Automapper_refseqTX
</pre></div>
</div>
<p>Example output of ds2015-149/ds2015-149_autoMapper_nr:</p>
<iframe src="_static/results/index.html" height="400px" width="100%"></iframe></div>
<div class="section" id="step-rps2tree">
<h3>Step <strong>Rps2tree</strong><a class="headerlink" href="#step-rps2tree" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div>This module use Rps2ecsv results of all sample to cut and group sequences based on identified domains and create OTUs, identity matrix, tree nexus files and png for each domains colored by SampleID.</div></blockquote>
<div class="highlight-bash"><div class="highlight"><pre><span></span>virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Rps2tree
</pre></div>
</div>
<iframe src="_static/rps2tree_global/index.html" height="400px" width="100%"></iframe></div>
<div class="section" id="step-getresults">
<h3>Step <strong>Getresults</strong><a class="headerlink" href="#step-getresults" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div>This module simply copy important results file to a result directory.</div></blockquote>
<div class="highlight-bash"><div class="highlight"><pre><span></span>virAnnot.py -m map.txt -s step.yaml -p parameters.yaml -n Getresults
</pre></div>
</div>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html#document-index">Table Of Contents</a></h3>
  <ul>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-prerequisite">Prerequisite</a><ul>
<li class="toctree-l2"><a class="reference internal" href="index.html#external-programs">External programs</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#external-databases">External databases</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#perl-external-libraries">Perl external libraries</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#perl-included-libraries">Perl included libraries</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#python-library">Python library</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#install">Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#database">Database</a><ul>
<li class="toctree-l2"><a class="reference internal" href="index.html#ncbi-taxonomy">NCBI Taxonomy</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#pfam-taxonomy">PFAM taxonomy</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#ncbi-blast-database">NCBI Blast database</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-parameter">Parameters files</a><ul>
<li class="toctree-l2"><a class="reference internal" href="index.html#parameters-yaml">parameters.yaml</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#step-yaml">step.yaml</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#map-txt">map.txt</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-modules">Modules description</a><ul>
<li class="toctree-l2"><a class="reference internal" href="index.html#readsoustraction">ReadSoustraction</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#demultiplex">Demultiplex</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#assembly">Assembly</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#map">Map</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#normalization">Normalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#diamond">Diamond</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#diamond2blast">Diamond2Blast</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#blast">Blast</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#blast2ecsv">Blast2ecsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#ecsv2excel">Ecsv2excel</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#ecsv2krona">Ecsv2krona</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#rps2ecsv">Rps2ecsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#rps2tree">Rps2tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#automapper">AutoMapper</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#blast2hits">Blast2hits</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-execution">Example execution</a><ul>
<li class="toctree-l2"><a class="reference internal" href="index.html#step-readsoustraction">Step <strong>ReadSoustraction</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#step-demultiplex">Step <strong>Demultiplex</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#step-demultiplexhtml">Step <strong>DemultiplexHtml</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#step-diamond">Step <strong>Diamond</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#step-assembly">Step <strong>Assembly</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#step-map">Step <strong>Map</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#step-blast">Step <strong>Blast</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#step-blast2ecsv">Step <strong>Blast2ecsv</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#step-rps2ecsv">Step <strong>Rps2ecsv</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#step-ecsv2excel">Step <strong>Ecsv2excel</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#step-ecsv2krona">Step <strong>Ecsv2krona</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#step-automapper">Step <strong>Automapper</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#step-rps2tree">Step <strong>Rps2tree</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#step-getresults">Step <strong>Getresults</strong></a></li>
</ul>
</li>
</ul>

        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="nav-item nav-item-0"><a href="index.html#document-index">VirAnnot 0.0.1 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2016, Sebastien Theil.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.3.
    </div>
  </body>
</html>