
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Modules description &#8212; VirAnnot 0.0.1 documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Example execution" href="execution.html" />
    <link rel="prev" title="Parameters files" href="parameter.html" /> 
  </head>
  <body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="execution.html" title="Example execution"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="parameter.html" title="Parameters files"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">VirAnnot 0.0.1 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="modules-description">
<h1>Modules description<a class="headerlink" href="#modules-description" title="Permalink to this headline">¶</a></h1>
<div class="section" id="readsoustraction">
<h2>ReadSoustraction<a class="headerlink" href="#readsoustraction" title="Permalink to this headline">¶</a></h2>
<p>ReadSoustraction module use bowtie2 to map reads against a reference. The output is directly piped to samtools for bam convertion and again piped to samtools to select unmapped paired reads (-f 12 -F 256). The pipe continues with bamtofastq tool to create two fastq files (r1 and r2).</p>
<div class="highlight-python3"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">_create_cmd</span> <span class="p">(</span><span class="bp">self</span><span class="p">):</span>
  <span class="n">cmd</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
  <span class="n">cmd</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;bin&#39;</span><span class="p">][</span><span class="s1">&#39;bowtie&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39; -p &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_cpu</span><span class="p">)</span>
  <span class="n">cmd</span> <span class="o">+=</span> <span class="s1">&#39; -x &#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span>
  <span class="n">cmd</span> <span class="o">+=</span> <span class="s1">&#39; -1 &#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">i1</span> <span class="o">+</span> <span class="s1">&#39; -2 &#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">i2</span> <span class="o">+</span> <span class="s1">&#39; | &#39;</span>
  <span class="n">cmd</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;bin&#39;</span><span class="p">][</span><span class="s1">&#39;samtools&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39; view -bS - &#39;</span>
  <span class="n">cmd</span> <span class="o">+=</span> <span class="s1">&#39; | &#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;bin&#39;</span><span class="p">][</span><span class="s1">&#39;samtools&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39; view -u -f 12 -F 256 - | &#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;bin&#39;</span><span class="p">][</span><span class="s1">&#39;bedtools&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39; bamtofastq -i - -fq &#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">o1</span> <span class="o">+</span> <span class="s1">&#39; -fq2 &#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">o2</span>
  <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
  <span class="k">return</span> <span class="n">cmd</span>
</pre></div>
</div>
<div class="section" id="options">
<h3>Options<a class="headerlink" href="#options" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">i1</span></code>: R1 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">i2</span></code>: R2 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">db</span></code>: Bowtie database.</li>
<li><code class="docutils literal"><span class="pre">o1</span></code>: R1 output fastq file.</li>
<li><code class="docutils literal"><span class="pre">o2</span></code>: R2 output fastq file.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL]</li>
<li><code class="docutils literal"><span class="pre">n_cpu</span></code>: [INT] Number of CPU to use.</li>
<li><code class="docutils literal"><span class="pre">iter</span></code>: [SampleID]</li>
</ul>
</div>
</div>
<div class="section" id="demultiplex">
<h2>Demultiplex<a class="headerlink" href="#demultiplex" title="Permalink to this headline">¶</a></h2>
<a class="reference internal image-reference" href="_images/demultiplex_0.png"><img alt="_images/demultiplex_0.png" src="_images/demultiplex_0.png" style="width: 750.0px; height: 470.4px;" /></a>
<p>Warning: This demultiplex procedure is specifc to our sequencing methods.</p>
<p>The demultiplex.pl script uses <a class="reference external" href="https://github.com/marcelm/cutadapt/">cutadapt</a> to demultiplex and trim step by step sequences used by the sequencing technology and the dsRNA extraction protocol.</p>
<p>Our extraction and sequencing protocol:</p>
<p>Informations about indexes, common sequences, and sample ids are stored in the map.txt file which must be tab delimited file.
The script produce lots of temporary file (*_step_*) that can be deleted at the end of the execution.
Each step produce 3 type of files:</p>
<dl class="docutils">
<dt>.info:</dt>
<dd>the matching information.</dd>
<dt>.log:</dt>
<dd>the execution information.</dd>
<dt>.out:</dt>
<dd>the fastq file.</dd>
</dl>
<div class="section" id="id1">
<h3>Options<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">i1</span></code>: R1 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">i2</span></code>: R2 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">adapters</span></code>: Fasta file of adapters.</li>
<li><code class="docutils literal"><span class="pre">middle</span></code>: Check for MIDs in middle of the reads and 1 : trim the reads or 2: exclude the read.</li>
<li><code class="docutils literal"><span class="pre">min_qual</span></code>: Trim the read if the quality is below this threshold.</li>
<li><code class="docutils literal"><span class="pre">polyA</span></code>: Trim poly-A tail.</li>
<li><code class="docutils literal"><span class="pre">min_len</span></code>: Exclude reads under this size threshold.</li>
<li><code class="docutils literal"><span class="pre">iter</span></code>: [SampleID]</li>
</ul>
</div>
<div class="section" id="step-01-5-index-search">
<h3>Step 01: 5’ index search<a class="headerlink" href="#step-01-5-index-search" title="Permalink to this headline">¶</a></h3>
<a class="reference internal image-reference" href="_images/demultiplex_1.png"><img alt="_images/demultiplex_1.png" src="_images/demultiplex_1.png" style="width: 1000.0px; height: 95.6px;" /></a>
<div class="highlight-perl"><div class="highlight"><pre><span></span><span class="n">_launchCutAdapt</span><span class="p">(</span><span class="nv">$self</span><span class="p">,</span><span class="nv">$files</span><span class="o">-&gt;</span><span class="p">{</span><span class="mi">1</span><span class="p">},</span> <span class="nv">$self</span><span class="o">-&gt;</span><span class="p">{</span><span class="nb">index</span><span class="p">},</span> <span class="nv">$tmp_file_prefix</span> <span class="o">.</span> <span class="s">&quot;_step.01_R1&quot;</span><span class="p">,</span><span class="s">&#39;d&#39;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;-g&#39;</span><span class="p">,</span><span class="s">&#39;0&#39;</span><span class="p">,</span><span class="s">&#39;1&#39;</span><span class="p">,</span><span class="s">&#39;0.8&#39;</span><span class="p">);</span>
</pre></div>
</div>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">'d','k'</span></code>: This step discard untrimmed reads, so reads that do not contain indexes are excluded, and keep trimmed reads.</li>
<li><code class="docutils literal"><span class="pre">'-g'</span></code>: search indexes in 5’.</li>
<li><code class="docutils literal"><span class="pre">'0'</span></code>: no errors allowed in the index sequence.</li>
<li><code class="docutils literal"><span class="pre">'1'</span></code>: search for one index in each read.</li>
<li><code class="docutils literal"><span class="pre">'0.8'</span></code>: 80% of the index length to be considered as a match.</li>
</ul>
</div>
<div class="section" id="step-02-5-common-sequence">
<h3>Step 02: 5’ common sequence<a class="headerlink" href="#step-02-5-common-sequence" title="Permalink to this headline">¶</a></h3>
<a class="reference internal image-reference" href="_images/demultiplex_2.png"><img alt="_images/demultiplex_2.png" src="_images/demultiplex_2.png" style="width: 1000.0px; height: 110.6px;" /></a>
<div class="highlight-perl"><div class="highlight"><pre><span></span><span class="n">_launchCutAdapt</span><span class="p">(</span><span class="nv">$self</span><span class="p">,</span><span class="nv">$files</span><span class="o">-&gt;</span><span class="p">{</span><span class="mi">1</span><span class="p">},</span><span class="nv">$self</span><span class="o">-&gt;</span><span class="p">{</span><span class="n">_common</span><span class="p">},</span><span class="nv">$tmp_file_prefix</span> <span class="o">.</span> <span class="s">&quot;_step.02_R1&quot;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;-g&#39;</span><span class="p">,</span><span class="s">&#39;0.1&#39;</span><span class="p">,</span><span class="nb">scalar</span><span class="p">(</span><span class="nb">keys</span><span class="p">(</span><span class="nv">%</span><span class="p">{</span><span class="nv">$self</span><span class="o">-&gt;</span><span class="p">{</span><span class="n">_common</span><span class="p">}})),</span><span class="s">&#39;0.7&#39;</span><span class="p">);</span>
</pre></div>
</div>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">'k','k'</span></code>: Keep reads that contains or not the common part.</li>
<li><code class="docutils literal"><span class="pre">'-g'</span></code>: search in 5’ part.</li>
<li><code class="docutils literal"><span class="pre">'0.1'</span></code>: 10% of sequencing errors.</li>
<li><code class="docutils literal"><span class="pre">scalar(keys(%{$self-&gt;{_common}}))</span></code>: will search as many common part as provided.</li>
<li><code class="docutils literal"><span class="pre">'0.7'</span></code>: 70% of the common part length to be considered as a match.</li>
</ul>
</div>
<div class="section" id="step-03-5-common-sequence-fragments">
<h3>Step 03: 5’ common sequence fragments<a class="headerlink" href="#step-03-5-common-sequence-fragments" title="Permalink to this headline">¶</a></h3>
<a class="reference internal image-reference" href="_images/demultiplex_3.png"><img alt="_images/demultiplex_3.png" src="_images/demultiplex_3.png" style="width: 1000.0px; height: 109.4px;" /></a>
<div class="highlight-perl"><div class="highlight"><pre><span></span><span class="n">_launchCutAdapt</span><span class="p">(</span><span class="nv">$self</span><span class="p">,</span><span class="nv">$files</span><span class="o">-&gt;</span><span class="p">{</span><span class="mi">1</span><span class="p">},</span><span class="nv">$self</span><span class="o">-&gt;</span><span class="p">{</span><span class="n">_common</span><span class="p">},</span><span class="nv">$tmp_file_prefix</span> <span class="o">.</span> <span class="s">&quot;_step.021_R1&quot;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;-g&#39;</span><span class="p">,</span><span class="s">&#39;0.2&#39;</span><span class="p">,</span><span class="nb">scalar</span><span class="p">(</span><span class="nb">keys</span><span class="p">(</span><span class="nv">%</span><span class="p">{</span><span class="nv">$self</span><span class="o">-&gt;</span><span class="p">{</span><span class="n">_common</span><span class="p">}})),</span><span class="s">&#39;0.5&#39;</span><span class="p">);</span>
</pre></div>
</div>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">'k','k'</span></code>: Keep reads that contains or not the common part.</li>
<li><code class="docutils literal"><span class="pre">'-g'</span></code>: search in 5’ part.</li>
<li><code class="docutils literal"><span class="pre">'0.2'</span></code>: 20% of sequencing errors.</li>
<li><code class="docutils literal"><span class="pre">scalar(keys(%{$self-&gt;{_common}}))</span></code>: will search as many common part as provided.</li>
<li><code class="docutils literal"><span class="pre">'0.5'</span></code>: 50% of the common part length to be considered as a match.</li>
</ul>
</div>
<div class="section" id="step-04-trimming-sequencing-adapters">
<h3>Step 04: Trimming sequencing adapters<a class="headerlink" href="#step-04-trimming-sequencing-adapters" title="Permalink to this headline">¶</a></h3>
<a class="reference internal image-reference" href="_images/demultiplex_4.png"><img alt="_images/demultiplex_4.png" src="_images/demultiplex_4.png" style="width: 1000.0px; height: 118.2px;" /></a>
<div class="highlight-perl"><div class="highlight"><pre><span></span><span class="n">_launchCutAdapt</span><span class="p">(</span><span class="nv">$self</span><span class="p">,</span><span class="nv">$files</span><span class="o">-&gt;</span><span class="p">{</span><span class="mi">1</span><span class="p">},</span><span class="nv">$self</span><span class="o">-&gt;</span><span class="p">{</span><span class="n">illuminaAdapter</span><span class="p">},</span><span class="nv">$tmp_file_prefix</span> <span class="o">.</span> <span class="s">&quot;_step.03_R1&quot;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;-b&#39;</span><span class="p">,</span><span class="s">&#39;0.2&#39;</span><span class="p">,</span><span class="nb">scalar</span><span class="p">(</span><span class="nb">keys</span><span class="p">(</span><span class="nv">%</span><span class="p">{</span><span class="nv">$self</span><span class="o">-&gt;</span><span class="p">{</span><span class="n">illuminaAdapter</span><span class="p">}})),</span><span class="s">&#39;0.6&#39;</span><span class="p">);</span>
</pre></div>
</div>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">'k','k'</span></code>: Keep reads that contains or not the common part.</li>
<li><code class="docutils literal"><span class="pre">'-b'</span></code>: search adapters anywhere in the read.</li>
<li><code class="docutils literal"><span class="pre">'0.2'</span></code>: 20% of sequencing errors.</li>
<li><code class="docutils literal"><span class="pre">scalar(keys(%{$self-&gt;{illuminaAdapter}}))</span></code>: will search as many adapters as provided.</li>
<li><code class="docutils literal"><span class="pre">'0.6'</span></code>: 60% of the adapters length to be considered as a match.</li>
</ul>
</div>
<div class="section" id="step-05-search-for-hybrid-reads">
<h3>Step 05: Search for hybrid reads<a class="headerlink" href="#step-05-search-for-hybrid-reads" title="Permalink to this headline">¶</a></h3>
<a class="reference internal image-reference" href="_images/demultiplex_5.png"><img alt="_images/demultiplex_5.png" src="_images/demultiplex_5.png" style="width: 1000.0px; height: 126.0px;" /></a>
<p>This step is really specific to our extraction method since very short DNA fragment can be link together during the aspecific adapters ligation step of the Illumina kits. This creating reads composed of two different PCR product. Thus our program search for index sequence in the middle of the read and trim it to keep the 5’ part or exlude the read.
The research is done both on provided indexes sequences and reverse complement of thoses sequences.</p>
<p><code class="docutils literal"><span class="pre">-middle</span> <span class="pre">[1|2]</span></code>   Search for common tag in the middle of the read. 1: trim the read. 2: exclude the read.</p>
<div class="highlight-perl"><div class="highlight"><pre><span></span><span class="n">_launchCutAdapt</span><span class="p">(</span><span class="nv">$self</span><span class="p">,</span><span class="nv">$files</span><span class="o">-&gt;</span><span class="p">{</span><span class="mi">1</span><span class="p">},</span><span class="nv">$h</span><span class="p">,</span><span class="nv">$tmp_file_prefix</span> <span class="o">.</span> <span class="s">&quot;_step.04_R1&quot;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;-b&#39;</span><span class="p">,</span><span class="s">&#39;0.1&#39;</span><span class="p">,</span><span class="s">&#39;1&#39;</span><span class="p">,</span><span class="s">&#39;0.5&#39;</span><span class="p">);</span>
</pre></div>
</div>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">'k','k'</span></code>: Keep reads that contains or not the index , or <code class="docutils literal"><span class="pre">'k','d'</span></code> if the <code class="docutils literal"><span class="pre">-middle</span></code> option is provided.</li>
<li><code class="docutils literal"><span class="pre">'-b'</span></code>: search adapters anywhere in the read.</li>
<li><code class="docutils literal"><span class="pre">'0.1'</span></code>: 20% of sequencing errors.</li>
<li><code class="docutils literal"><span class="pre">'1'</span></code>: search for one index in each read.</li>
<li><code class="docutils literal"><span class="pre">'0.5'</span></code>: 50% of the adapters length to be considered as a match.</li>
</ul>
</div>
<div class="section" id="step-06-search-for-polya-optional">
<h3>Step 06: Search for polyA (optional)<a class="headerlink" href="#step-06-search-for-polya-optional" title="Permalink to this headline">¶</a></h3>
<a class="reference internal image-reference" href="_images/demultiplex_6.png"><img alt="_images/demultiplex_6.png" src="_images/demultiplex_6.png" style="width: 750.0px; height: 150.0px;" /></a>
<p>In Illumina technology, if the sequencing matrix is too short compared to the sequencing length, the sequencing machine adds a bunch of A’s and then radom sequence.</p>
<div class="highlight-perl"><div class="highlight"><pre><span></span><span class="n">_launchCutAdapt</span><span class="p">(</span><span class="nv">$self</span><span class="p">,</span><span class="nv">$files</span><span class="o">-&gt;</span><span class="p">{</span><span class="mi">1</span><span class="p">},</span> <span class="nv">$h</span><span class="p">,</span> <span class="nv">$tmp_file_prefix</span> <span class="o">.</span> <span class="s">&quot;_step.05_R1&quot;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;k&#39;</span><span class="p">,</span><span class="s">&#39;-a&#39;</span><span class="p">,</span><span class="s">&#39;0&#39;</span><span class="p">,</span><span class="s">&#39;1&#39;</span><span class="p">,</span><span class="s">&#39;0.8&#39;</span><span class="p">);</span>
</pre></div>
</div>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">'k','k'</span></code>: Keep reads that contains or not the index , or <code class="docutils literal"><span class="pre">'k','d'</span></code> if the <code class="docutils literal"><span class="pre">-middle</span></code> option is provided.</li>
<li><code class="docutils literal"><span class="pre">'-a'</span></code>: search in the 3’ end.</li>
<li><code class="docutils literal"><span class="pre">'0'</span></code>: no sequencing errors.</li>
<li><code class="docutils literal"><span class="pre">'1'</span></code>: search for one index in each read.</li>
<li><code class="docutils literal"><span class="pre">'0.8'</span></code>: 80% of the polyA length to be considered as a match.</li>
</ul>
</div>
</div>
<div class="section" id="assembly">
<h2>Assembly<a class="headerlink" href="#assembly" title="Permalink to this headline">¶</a></h2>
<p>This module can launch two assemblers, <a class="reference external" href="https://github.com/loneknightpy/idba/">IDBA</a> and <a class="reference external" href="http://bioinf.spbau.ru/en/spades3.7/">MetaSpades</a>.</p>
<p>Foreach assembler, the module convert reads files to the proper format, launch the assembly in a separate directory, rename scaffolds identifier and move results file to the sample root directory.</p>
<div class="section" id="id2">
<h3>Options<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<div class="section" id="map">
<h2>Map<a class="headerlink" href="#map" title="Permalink to this headline">¶</a></h2>
<p>This module uses bowtie2, samtools and readPerContig.pl script to map reads back on the assembly and count for each scaffold the number of reads aligned resulting a simple two column file scaffoldID and nb_reads used by other modules.</p>
<div class="section" id="id3">
<h3>Options<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">contigs</span></code>: fasta file of contigs to map reads on. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">i1</span></code>: R1 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">i2</span></code>: R2 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">ising</span></code>: singletons fastq file</li>
<li><code class="docutils literal"><span class="pre">n_cpu</span></code>: [INT] number of CPU to use.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL] use SGE scheduler.</li>
<li><code class="docutils literal"><span class="pre">bam</span></code>: BAM file name.</li>
<li><code class="docutils literal"><span class="pre">rn</span></code>: output file name.</li>
</ul>
</div>
</div>
<div class="section" id="normalization">
<h2>Normalization<a class="headerlink" href="#normalization" title="Permalink to this headline">¶</a></h2>
<p>This module randomly select NUM reads from paired-files.</p>
<div class="section" id="id4">
<h3>Options<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">i1</span></code>: R1 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">i2</span></code>: R2 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">o1</span></code>: Output R1 normalized file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">o2</span></code>: Output R2 normalized file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">num</span></code>: [INT] Number of reads to randomly select. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">iter</span></code>: Iteration on [sample, library].</li>
<li><code class="docutils literal"><span class="pre">n_cpu</span></code>: [INT] number of CPU to use.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL] use SGE scheduler.</li>
</ul>
</div>
</div>
<div class="section" id="diamond">
<h2>Diamond<a class="headerlink" href="#diamond" title="Permalink to this headline">¶</a></h2>
<p>This module launches Diamond similarity search on reads and produce an XML file simalar to what Blast does so it can be treated by the Blast2ecsv module and so on.</p>
<div class="section" id="id5">
<h3>Options<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">i1</span></code>: R1 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">i2</span></code>: R2 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">db</span></code>: Values are defined in the parameters.yaml file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">ising</span></code>: singletons fastq file</li>
<li><code class="docutils literal"><span class="pre">n_cpu</span></code>: [INT] number of CPU to use.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL] use SGE scheduler.</li>
<li><code class="docutils literal"><span class="pre">sensitive</span></code>: [BOOL]</li>
<li><code class="docutils literal"><span class="pre">more_sensitive</span></code>: [BOOL]</li>
<li><code class="docutils literal"><span class="pre">out</span></code>: XML output file</li>
<li><code class="docutils literal"><span class="pre">score</span></code>: Report matches above this score.</li>
<li><code class="docutils literal"><span class="pre">max_target_seqs</span></code>: Maximum match per query sequences.</li>
<li><code class="docutils literal"><span class="pre">evalue</span></code>: Min e-value.</li>
<li><code class="docutils literal"><span class="pre">identity</span></code>: Report matches above this identity percent. 0 &gt; X &gt; 100.</li>
<li><code class="docutils literal"><span class="pre">qov</span></code>: Query overlap.</li>
<li><code class="docutils literal"><span class="pre">hov</span></code>: Hit overlap.</li>
</ul>
</div>
</div>
<div class="section" id="diamond2blast">
<h2>Diamond2Blast<a class="headerlink" href="#diamond2blast" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id6">
<h3>Options<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">i</span></code>: CSV file with DIAMOND results. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">contigs</span></code>: Fasta file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">out</span></code>: XML output file.</li>
<li><code class="docutils literal"><span class="pre">type</span></code>: Blast type. [‘tblastx’,’blastx’,’blastn’,’blastp’,’rpstblastn’]. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">db</span></code>: Values are defined in the parameters.yaml file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">evalue</span></code>: Min e-value.</li>
<li><code class="docutils literal"><span class="pre">server</span></code>: [‘enki’,’genologin’,’avakas’] Values are defined in the parameters.yaml file.</li>
<li><code class="docutils literal"><span class="pre">n_cpu</span></code>: [INT] number of CPU to use.</li>
<li><code class="docutils literal"><span class="pre">tc</span></code>: Number of task launched at the same time on SGE.</li>
<li><code class="docutils literal"><span class="pre">num_chunk</span></code>: Number of chunks to split the original fasta file for parallel execution.</li>
<li><code class="docutils literal"><span class="pre">max_target_seqs</span></code>: Maximum match per query sequences.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL] use SGE scheduler.</li>
</ul>
</div>
</div>
<div class="section" id="blast">
<h2>Blast<a class="headerlink" href="#blast" title="Permalink to this headline">¶</a></h2>
<p>This module launches all type of Blast on local machine or distant servers. This module has been developped for our own local machines and servers, but it can be easly modified to fit your needs.</p>
<p>This module mainly depends on the parameters.yaml file and the blast_launch.py script which has to be present on the server you want to use and modified to fit your server configuration.</p>
<div class="section" id="id7">
<h3>Options<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">contigs</span></code>: Fasta file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">db</span></code>: Values are defined in the parameters.yaml file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">type</span></code>: Blast type. [‘tblastx’,’blastx’,’blastn’,’blastp’,’rpstblastn’]. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">n_cpu</span></code>: [INT] number of CPU to use.</li>
<li><code class="docutils literal"><span class="pre">tc</span></code>: Number of task launched at the same time on SGE. (Experimental, works on Genotoul)</li>
<li><code class="docutils literal"><span class="pre">max_target_seqs</span></code>: Maximum match per query sequences.</li>
<li><code class="docutils literal"><span class="pre">num_chunk</span></code>: Number of chunks to split the original fasta file for parallel execution.</li>
<li><code class="docutils literal"><span class="pre">out</span></code>: Output file name.</li>
<li><code class="docutils literal"><span class="pre">server</span></code>: [‘enki’,’genologin’,’avakas’] Values are defined in the parameters.yaml file.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL] use SGE scheduler.</li>
</ul>
<p>This module is able to launch Blast instance on distant servers if the database and the blast_launch.py script is present on the server. Then you have to edit the parameters.yaml file to fit your configuration. The script has been developped to use two computer cluster, Avakas (PBS + Torque) and Genotoul (SGE) but each cluster has its own configuration so you may have to modify this script to adapt it to your configuration.</p>
</div>
</div>
<div class="section" id="blast2ecsv">
<h2>Blast2ecsv<a class="headerlink" href="#blast2ecsv" title="Permalink to this headline">¶</a></h2>
<p>This module parse Blast xml outputs, filter matches on different criteria and link Accession number to NCBI taxonomy.</p>
<div class="section" id="id8">
<h3>Options<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">b</span></code>: Blast file.</li>
<li><code class="docutils literal"><span class="pre">if</span></code>: Input format [‘xml’,’m8’]</li>
<li><code class="docutils literal"><span class="pre">out</span></code>: Output file name.</li>
<li><code class="docutils literal"><span class="pre">evalue</span></code>: Min e-value.</li>
<li><code class="docutils literal"><span class="pre">fhit</span></code>: Only report first hit.</li>
<li><code class="docutils literal"><span class="pre">fhsp</span></code>: Only report first hsp.</li>
<li><code class="docutils literal"><span class="pre">pm</span></code>:</li>
<li><code class="docutils literal"><span class="pre">r</span></code>: Reduced taxonomy. Report only 5 consistent rank.</li>
<li><code class="docutils literal"><span class="pre">vs</span></code>: Only report sequences when match is virus or viroids.</li>
<li><code class="docutils literal"><span class="pre">rn</span></code>: Read number. File created by the Map module.</li>
<li><code class="docutils literal"><span class="pre">type</span></code>: Blast type. [‘tblastx’,’blastx’,’blastn’,’blastp’,’rpstblastn’]</li>
<li><code class="docutils literal"><span class="pre">score</span></code>: Report matches above this score.</li>
<li><code class="docutils literal"><span class="pre">identity</span></code>: Report matches above this identity percent. 0 &gt; X &gt; 100.</li>
<li><code class="docutils literal"><span class="pre">qov</span></code>: Query overlap.</li>
<li><code class="docutils literal"><span class="pre">hov</span></code>: Hit overlap.</li>
<li><code class="docutils literal"><span class="pre">pd</span></code>: Parse description. Useful when the query ID is stored in the dscription field in the XML file.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL] use SGE scheduler.</li>
</ul>
</div>
</div>
<div class="section" id="ecsv2excel">
<h2>Ecsv2excel<a class="headerlink" href="#ecsv2excel" title="Permalink to this headline">¶</a></h2>
<p>This module aggregates multiple ecsv file to create a colored XLSX file.
It launches the ecsv2krona.pl script.</p>
<div class="section" id="id9">
<h3>Options<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">b[INT]</span></code>: CSV Blast file from 1 to 10.</li>
<li><code class="docutils literal"><span class="pre">out</span></code>: Outpuy file name.</li>
<li><code class="docutils literal"><span class="pre">r</span></code>: RPSBLAST csv file.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL] use SGE scheduler.</li>
</ul>
</div>
</div>
<div class="section" id="ecsv2krona">
<h2>Ecsv2krona<a class="headerlink" href="#ecsv2krona" title="Permalink to this headline">¶</a></h2>
<p>This module launch the ecsv2krona.pl script. It will aggregate multiple ecsv file into one Krona html file.</p>
<div class="section" id="id10">
<h3>Options<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">b</span></code>: [INT] CSV Blast file.</li>
<li><code class="docutils literal"><span class="pre">id</span></code>: [INT] ID wanted corresponding to the Blast file.</li>
<li><code class="docutils literal"><span class="pre">x</span></code>: [INT] XML Blast file. If used, this file will be split by species and link in the Krona file.</li>
<li><code class="docutils literal"><span class="pre">out</span></code>: Output file name.</li>
<li><code class="docutils literal"><span class="pre">data</span></code>: [‘both’,’reads’,’contigs’,’none’]</li>
<li><code class="docutils literal"><span class="pre">r</span></code>: Use reduced taxonomy.</li>
<li><code class="docutils literal"><span class="pre">c</span></code>: [‘identity’,’taxid’,’none’]</li>
<li><code class="docutils literal"><span class="pre">iter</span></code>: [‘global’]</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL] use SGE scheduler.</li>
</ul>
</div>
</div>
<div class="section" id="rps2ecsv">
<h2>Rps2ecsv<a class="headerlink" href="#rps2ecsv" title="Permalink to this headline">¶</a></h2>
<p>This module launch rps2ecsv.pl script for each sample.
This module parse XML files from rpsblast and create csv file as a result.</p>
<div class="section" id="id11">
<h3>Options<a class="headerlink" href="#id11" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">b</span></code>: RPSBLAST XML file.</li>
<li><code class="docutils literal"><span class="pre">contigs</span></code>: Fasta file.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL] use SGE scheduler.</li>
<li><code class="docutils literal"><span class="pre">out</span></code>: Output file name.</li>
<li><code class="docutils literal"><span class="pre">evalue</span></code>: e-value threshold.</li>
</ul>
</div>
</div>
<div class="section" id="rps2tree">
<h2>Rps2tree<a class="headerlink" href="#rps2tree" title="Permalink to this headline">¶</a></h2>
<p>This module launch rps2tree.pl script for all sample provided.
This module generates Operational Taxonomic Unit (OTU) based on RPS-Blast results.
For each CDD motifs, contigs are clustered together based on matrix distance.</p>
<div class="section" id="id12">
<h3>Options<a class="headerlink" href="#id12" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">pfam</span></code>: CSV file from Rps2ecsv.</li>
<li><code class="docutils literal"><span class="pre">contigs</span></code>: Fasta file.</li>
<li><code class="docutils literal"><span class="pre">ecsv</span></code>: CSV file from Blast2ecsv.</li>
<li><code class="docutils literal"><span class="pre">out</span></code>: Output file name.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL]</li>
<li><code class="docutils literal"><span class="pre">viral_portion</span></code>: Minimum percentage of viral sequence in a domain to be selected.</li>
<li><code class="docutils literal"><span class="pre">min_prot</span></code>: Minimum protein length to be included in a tree.</li>
<li><code class="docutils literal"><span class="pre">perc</span></code>: Percentage of identity. Threshold set to define OTU.</li>
<li><code class="docutils literal"><span class="pre">iter</span></code>: [‘global’]</li>
</ul>
</div>
</div>
<div class="section" id="automapper">
<h2>AutoMapper<a class="headerlink" href="#automapper" title="Permalink to this headline">¶</a></h2>
<p>This module launch autoMapper.pl script for every sample.</p>
<div class="section" id="id13">
<h3>Options<a class="headerlink" href="#id13" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">contigs</span></code>: Fasta file.</li>
<li><code class="docutils literal"><span class="pre">ecsv</span></code>: CSV file from Blast2ecsv.</li>
<li><code class="docutils literal"><span class="pre">i1</span></code>: R1 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">i2</span></code>: R2 fastq file. [mandatory]</li>
<li><code class="docutils literal"><span class="pre">out</span></code>: Output folder.</li>
<li><code class="docutils literal"><span class="pre">sge</span></code>: [BOOL]</li>
<li><code class="docutils literal"><span class="pre">ref</span></code>: Blast database name.</li>
</ul>
</div>
</div>
<div class="section" id="blast2hits">
<h2>Blast2hits<a class="headerlink" href="#blast2hits" title="Permalink to this headline">¶</a></h2>
<p>This takes multiple CSV Blast file from Blast2ecsv and draw histograms for by taxonomy.</p>
<div class="section" id="id14">
<h3>Options<a class="headerlink" href="#id14" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">b[INT]</span></code> CSV Blast file from Blast2ecsv</li>
<li><code class="docutils literal"><span class="pre">id[INT]</span></code> ID associated with the Blast file.</li>
<li><code class="docutils literal"><span class="pre">iter</span></code> [global]</li>
<li><code class="docutils literal"><span class="pre">sge</span></code> [BOOL]</li>
<li><code class="docutils literal"><span class="pre">out</span></code> Output file name.</li>
</ul>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Modules description</a><ul>
<li><a class="reference internal" href="#readsoustraction">ReadSoustraction</a><ul>
<li><a class="reference internal" href="#options">Options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#demultiplex">Demultiplex</a><ul>
<li><a class="reference internal" href="#id1">Options</a></li>
<li><a class="reference internal" href="#step-01-5-index-search">Step 01: 5’ index search</a></li>
<li><a class="reference internal" href="#step-02-5-common-sequence">Step 02: 5’ common sequence</a></li>
<li><a class="reference internal" href="#step-03-5-common-sequence-fragments">Step 03: 5’ common sequence fragments</a></li>
<li><a class="reference internal" href="#step-04-trimming-sequencing-adapters">Step 04: Trimming sequencing adapters</a></li>
<li><a class="reference internal" href="#step-05-search-for-hybrid-reads">Step 05: Search for hybrid reads</a></li>
<li><a class="reference internal" href="#step-06-search-for-polya-optional">Step 06: Search for polyA (optional)</a></li>
</ul>
</li>
<li><a class="reference internal" href="#assembly">Assembly</a><ul>
<li><a class="reference internal" href="#id2">Options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#map">Map</a><ul>
<li><a class="reference internal" href="#id3">Options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#normalization">Normalization</a><ul>
<li><a class="reference internal" href="#id4">Options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#diamond">Diamond</a><ul>
<li><a class="reference internal" href="#id5">Options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#diamond2blast">Diamond2Blast</a><ul>
<li><a class="reference internal" href="#id6">Options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#blast">Blast</a><ul>
<li><a class="reference internal" href="#id7">Options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#blast2ecsv">Blast2ecsv</a><ul>
<li><a class="reference internal" href="#id8">Options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#ecsv2excel">Ecsv2excel</a><ul>
<li><a class="reference internal" href="#id9">Options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#ecsv2krona">Ecsv2krona</a><ul>
<li><a class="reference internal" href="#id10">Options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#rps2ecsv">Rps2ecsv</a><ul>
<li><a class="reference internal" href="#id11">Options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#rps2tree">Rps2tree</a><ul>
<li><a class="reference internal" href="#id12">Options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#automapper">AutoMapper</a><ul>
<li><a class="reference internal" href="#id13">Options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#blast2hits">Blast2hits</a><ul>
<li><a class="reference internal" href="#id14">Options</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="parameter.html"
                        title="previous chapter">Parameters files</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="execution.html"
                        title="next chapter">Example execution</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/modules.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="execution.html" title="Example execution"
             >next</a> |</li>
        <li class="right" >
          <a href="parameter.html" title="Parameters files"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">VirAnnot 0.0.1 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2016, Sebastien Theil.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.3.
    </div>
  </body>
</html>