<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Parameters files &mdash; VirAnnot 0.0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="VirAnnot 0.0.1 documentation" href="index.html" />
    <link rel="next" title="Modules description" href="modules.html" />
    <link rel="prev" title="Prerequisite" href="prerequisite.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="modules.html" title="Modules description"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="prerequisite.html" title="Prerequisite"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">VirAnnot 0.0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="parameters-files">
<h1>Parameters files<a class="headerlink" href="#parameters-files" title="Permalink to this headline">¶</a></h1>
<div class="section" id="parameters-yaml">
<h2>parameters.yaml<a class="headerlink" href="#parameters-yaml" title="Permalink to this headline">¶</a></h2>
<p>Defines paths for both local and remote binaries and databases. A template is provided in the <code class="docutils literal"><span class="pre">examples</span></code> directory.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>ReadSoustraction:
  db:
    vitis: &#39;/media/data/db/ncbi/vitis/vitis&#39;
    phiX:  &#39;/media/data/db/ncbi/phiX/phiX174&#39;
bin:
  bowtie: &#39;/usr/local/bin/bowtie2&#39;
  samtools: &#39;/usr/bin/samtools&#39;
  bedtools: &#39;/usr/bin/bedtools&#39;
  prinseq: &#39;/usr/local/bin/prinseq-lite.pl&#39;
servers:
  genotoul:
    adress: &#39;genotoul.toulouse.inra.fr&#39;
    username: &#39;stheil&#39;
    db:
      nr: &#39;/bank/blastdb/nr&#39;
      nt: &#39;/bank/blastdb/nt&#39;
      refseq_vir_nucl: &#39;/save/stheil/db/refseq_vir/viral.genomic.fna&#39;
      refseq_vir_prot: &#39;/save/stheil/db/refseq_vir/viral.protein.faa&#39;
      pfam: &#39;/home/stheil/save/db/pfam/pfam_viruses_rpsdb&#39;
      all_vir_nucl: &#39;/home/stheil/save/db/all_vir/all_vir_nucl.fna&#39;
      all_vir_prot: &#39;/home/stheil/save/db/all_vir/all_vir_prot.faa&#39;
    scratch: &#39;/work/stheil&#39;
    bin:
      blastx: &#39;blastx+&#39;
      blastn: &#39;blastn+&#39;
  avakas:
    adress: &#39;avakas.mcia.univ-bordeaux.fr&#39;
    username: &#39;stheil&#39;
    db:
      nr: &#39;/home/stheil/db/nr/nr&#39;
      nt: &#39;/home/stheil/db/nt/nt&#39;
      all_vir_nucl: &#39;/home/stheil/scratch/db/all_vir/all_vir_nucl.fna&#39;
      all_vir_prot: &#39;/home/stheil/scratch/db/all_vir/all_vir_prot.faa&#39;
      refseq_vir_nucl: &#39;/home/stheil/scratch/db/refseq_vir/viral.genomic.fna&#39;
      refseq_vir_prot: &#39;/home/stheil/scratch/db/refseq_vir/viral.protein.faa&#39;
      pfam: &#39;/home/stheil/db/pfam/pfam_viruses_rpsdb&#39;
    scratch: &#39;/scratch/stheil&#39;
    bin:
      blastx: &#39;blastx&#39;
      blastn: &#39;blastn&#39;
AutoMapper:
  db:
    nt: /media/data/db/ncbi/nt/nt
    refseq_vir_nucl: /media/data/db/ncbi/refseq_vir/viral.genomic.fna
    all_vir_nucl: /media/data/db/ncbi/all_vir/all_vir_nucl.fna
</pre></div>
</div>
</div>
<div class="section" id="step-yaml">
<h2>step.yaml<a class="headerlink" href="#step-yaml" title="Permalink to this headline">¶</a></h2>
<p>Defines the steps that the pipeline will execute. A template is provided in the <code class="docutils literal"><span class="pre">/examples</span></code> directory.</p>
<p>Step names correspond to a python module that will launch the step. Step names are split based on the &#8216;_&#8217; character so you can launch multiple instance. For example you might want to launch blastx and blastn, so step names could be &#8216;Blast_N&#8217; and &#8216;Blast_X&#8217;. What is after the underscore do not matters, it is just used to differanciate the two steps.</p>
<p>Special words in bracket are used as substitution string.
- (file), (file1) and (file2)
- (SampleID)
- (library)</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>ReadSoustraction_phiX:
  i1: (file1)
  i2: (file2)
  db: phiX
  o1: (library)_phiX.r1.fq
  o2: (library)_phiX.r2.fq
  sge: True
  n_cpu: 5
  iter: library
Demultiplex:
  i1: (library)_phiX.r1.fq
  i2: (library)_phiX.r2.fq
  adapters: adapters.fna
  middle: 1
  min_qual: 20
  polyA: True
  min_len: 70
  iter: library
  sge: True
DemultiplexHtml:
  csv: (library)_demultiplex.stats.csv
  id: (library)
  out: stat_demultiplex
  iter: global
  sge: True
drVM:
  i1: (SampleID)_truePairs_r1.fq
  i2: (SampleID)_truePairs_r2.fq
  n_cpu: 20
  identity: 70
  min_len: 300
  sge: True
Assembly_idba:
  prog: idba
  n_cpu: 5
  i1: (SampleID)_truePairs_r1.fq
  i2: (SampleID)_truePairs_r2.fq
  out: (SampleID)_idba.scaffold.fa
  sge: True
Assembly_spades:
  prog: spades
  n_cpu: 5
  i1: (SampleID)_truePairs_r1.fq
  i2: (SampleID)_truePairs_r2.fq
  out: (SampleID)_spades.scaffold.fa
  sge: True
Map_idba:
  contigs: (SampleID)_idba.scaffold.fa
  i1: (SampleID)_truePairs_r1.fq
  i2: (SampleID)_truePairs_r2.fq
  bam: (SampleID)_idba.scaffold.bam
  rn: (SampleID)_idba.scaffold.rn
  sge: True
  n_cpu: 16
Map_spades:
  contigs: (SampleID)_spades.scaffold.fa
  i1: (SampleID)_truePairs_r1.fq
  i2: (SampleID)_truePairs_r2.fq
  bam: (SampleID)_spades.scaffold.bam
  rn: (SampleID)_spades.scaffold.rn
  sge: True
  n_cpu: 16
Diamond:
  i1: (SampleID)_truePairs_r1.fq
  i2: (SampleID)_truePairs_r2.fq
  n_cpu: 10
  sge: True
  score: 50
  evalue: 0.0001
  qov: 50
  hov: 5
  db: all_vir_prot
Blast2ecsv_dmd:
  evalue: 0.01
  fhit: True
  pm: global
  if: xml
  r: True
  b: (SampleID)_dmd.xml
  out: (SampleID)_dmd.allVirProt.csv
  sge: True
  type: BLASTX
  pd: True
Ecsv2krona_dmd:
  id1: (SampleID)
  b1: (SampleID)_dmd.allVirProt.csv
  outdir: krona_diamond
  out: global_krona_dmd.html
  data: contig
  r: True
  c: identity
  iter: global
  sge: True
Blast_allvirTX:
  type: tblastx
  contigs: (SampleID)_idba.scaffold.fa
  db: all_vir_nucl
  out: (SampleID)_idba.scaffold.tbltx.all_vir.xml
  evalue: 0.0001
  server: genotoul
  n_cpu: 8
  sge: True
  num_chunk: 1000
  tc: 50
Blast_nr:
  type: blastx
  contigs: (SampleID)_idba.scaffold.fa
  db: nr
  out: (SampleID)_idba.scaffold.bltx.nr.xml
  evalue: 0.0001
  server: genotoul
  n_cpu: 8
  tc: 50
  num_chunk: 1000
  max_target_seqs: 1
  sge: True
Blast_refvirTX:
  type: tblastx
  contigs: (SampleID)_idba.scaffold.fa
  db: refseq_vir_nucl
  out: (SampleID)_idba.scaffold.tbltx.refseq_vir.xml
  evalue: 0.0001
  server: genotoul
  n_cpu: 8
  tc: 50
  num_chunk: 1000
  sge: True
Blast_RPS:
  type: rpstblastn
  contigs: (SampleID)_idba.scaffold.fa
  db: pfam
  evalue: 0.0001
  out: (SampleID)_idba.scaffold.rps.pfam.xml
  server: genotoul
  n_cpu: 8
  sge: True
Blast2ecsv_allvirTX:
  contigs: (SampleID)_idba.scaffold.fa
  evalue: 0.001
  fhit: True
  pm: global
  if: xml
  rn: (SampleID)_idba.scaffold.rn
  r: True
  b: (SampleID)_idba.scaffold.tbltx.all_vir.xml
  vs: True
  out: (SampleID)_idba.scaffold.tbltx.all_vir.csv
  sge: True
  type: TBLASTX
  score: 50
  qov: 20
Blast2ecsv_refvirTX:
  contigs: (SampleID)_idba.scaffold.fa
  evalue: 0.0001
  fhit: True
  pm: global
  if: xml
  rn: (SampleID)_idba.scaffold.rn
  r: True
  b: (SampleID)_idba.scaffold.tbltx.refseq_vir.xml
  vs: True
  out: (SampleID)_idba.scaffold.tbltx.refseq_vir.csv
  sge: True
  type: TBLASTX
  score: 50
  qov: 50
  hov: 5
Blast2ecsv_nr:
  contigs: (SampleID)_idba.scaffold.fa
  evalue: 0.001
  fhit: True
  pm: global
  if: xml
  rn: (SampleID)_idba.scaffold.rn
  r: True
  b: (SampleID)_idba.scaffold.bltx.nr.xml
  vs: True
  out: (SampleID)_idba.scaffold.bltx.nr.csv
  sge: True
  type: BLASTX
  score: 50
  qov: 5
  hov: 5
Rps2ecsv:
  b: (SampleID)_idba.scaffold.rps.pfam.xml
  out: (SampleID)_idba.scaffold.rps.pfam.csv
  evalue: 0.0001
  sge: True
Ecsv2excel:
  b1: (SampleID)_idba.scaffold.tbltx.refseq_vir.csv
  b2: (SampleID)_idba.scaffold.tbltx.all_vir.csv
  b3: (SampleID)_idba.scaffold.bltx.nr.csv
  r:  (SampleID)_idba.scaffold.rps.pfam.csv
  out:  (SampleID)_idba.scaffold.xlsx
  sge: True
Blast2hist:
  id1: (SampleID)_refseq_tbltx
  b1: (SampleID)_idba.scaffold.tbltx.refseq_vir.csv
  id2: (SampleID)_allvir_tbltx
  b2: (SampleID)_idba.scaffold.tbltx.all_vir.csv
  id3: (SampleID)_nr_bltx
  b3: (SampleID)_idba.scaffold.bltx.nr.csv
  id4: (SampleID)_dmd
  b4: (SampleID)_dmd.allVirProt.csv
  iter: global
  sge: True
  out: blast_hist
Ecsv2krona:
  id1: (SampleID)_refseq_tbltx
  b1: (SampleID)_idba.scaffold.tbltx.refseq_vir.csv
  x1: (SampleID)_idba.scaffold.tbltx.refseq_vir.xml
  id2: (SampleID)_allvir_tbltx
  b2: (SampleID)_idba.scaffold.tbltx.all_vir.csv
  x2: (SampleID)_idba.scaffold.tbltx.all_vir.xml
  id3: (SampleID)_nr_bltx
  b3: (SampleID)_idba.scaffold.bltx.nr.csv
  x3: (SampleID)_idba.scaffold.bltx.nr.xml
  outdir: krona_blast
  out: blast.global.krona.html
  data: both
  r: True
  c: identity
  iter: global
  sge: True
Automapper_nr:
  contigs: (SampleID)_idba.scaffold.fa
  ecsv: (SampleID)_idba.scaffold.bltx.nr.csv
  i1: (SampleID)_truePairs_r1.fq
  i2: (SampleID)_truePairs_r2.fq
  out: (SampleID)_autoMapper_nr
  sge: True
  ref: nt
Automapper_allvirTX:
  contigs: (SampleID)_idba.scaffold.fa
  ecsv: (SampleID)_idba.scaffold.tbltx.all_vir.csv
  i1: (SampleID)_truePairs_r1.fq
  i2: (SampleID)_truePairs_r2.fq
  out: (SampleID)_autoMapper_allvir
  sge: True
  ref: all_vir_nucl
Automapper_refseqTX:
  contigs: (SampleID)_idba.scaffold.fa
  ecsv: (SampleID)_idba.scaffold.tbltx.refseq_vir.csv
  i1: (SampleID)_truePairs_r1.fq
  i2: (SampleID)_truePairs_r2.fq
  out: (SampleID)_autoMapper_refseq
  sge: True
  ref: refseq_vir_nucl
Rps2tree:
  pfam: (SampleID)_idba.scaffold.rps.pfam.csv
  contigs: (SampleID)_idba.scaffold.fa
  ecsv: (SampleID)_idba.scaffold.bltx.nr.csv
  id: (SampleID)
  out: rps2tree_global
  min_prot: 100
  viral_portion: 0.3
  iter: global
  sge: True
Getresults:
  global_dir1: rps2tree_global
  global_dir2: krona_blast
  global_dir3: krona_diamond
  global_dir4: blast_hist
  global_dir5: stat_demultiplex
  sample_dir1: (SampleID)_autoMapper_nr
  sample_dir2: (SampleID)_autoMapper_refseq
  sample_dir3r: (SampleID)_autoMapper_allvir
  sample_file1: (SampleID)_idba.scaffold.xlsx
  sample_file2: (SampleID)_idba.scaffold.fa
  sample_file3: (SampleID)_spades.scaffold.fa
  sample_file4: (SampleID)_truePairs_r1.fq
  sample_file5: (SampleID)_truePairs_r2.fq
  out: results
</pre></div>
</div>
</div>
<div class="section" id="map-txt">
<span id="id1"></span><h2>map.txt<a class="headerlink" href="#map-txt" title="Permalink to this headline">¶</a></h2>
<p>The map file describe the experiment. It is a tabulated file with the first line containing headers starting with &#8216;#&#8217;. It must contain at least two column: SampleID and file.
A template is provided in the <code class="docutils literal"><span class="pre">examples</span></code> directory.
This is a minimum map.txt file:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>#SampleID	mid	common	file1	file2	library
ds2016-121	AACCGCAA	TGTGTTGGGTGTGTTTGG	Lib1_phiX.R1.fastq	Lib1_phiX.R2.fastq	lib1
ds2016-132	AACTAGTA	TGTGTTGGGTGTGTTTGG	Lib1_phiX.R1.fastq	Lib1_phiX.R2.fastq	lib1
ds2016-122	AGGCGCCT	TGTGTTGGGTGTGTTTGG	Lib2_phiX.R1.fastq	Lib2_phiX.R2.fastq	lib2
ds2016-133	ATTAGCTA	TGTGTTGGGTGTGTTTGG	Lib2_phiX.R1.fastq	Lib2_phiX.R2.fastq	lib2
ds2016-123	CAAGAGTT	TGTGTTGGGTGTGTTTGG	Lib3_phiX.R1.fastq	Lib3_phiX.R2.fastq	lib3
ds2016-55	CAAGCAGG	TGTGTTGGGTGTGTTTGG	Lib3_phiX.R1.fastq	Lib3_phiX.R2.fastq	lib3
ds2016-124	CCAACCAT	TGTGTTGGGTGTGTTTGG	Lib4_phiX.R1.fastq	Lib4_phiX.R2.fastq	lib4
ds2016-56	CGATAGAG	TGTGTTGGGTGTGTTTGG	Lib4_phiX.R1.fastq	Lib4_phiX.R2.fastq	lib4
ds2016-125	GCTCTACC	TGTGTTGGGTGTGTTTGG	Lib5_phiX.R1.fastq	Lib5_phiX.R2.fastq	lib5
ds2016-57	GCTGCGGT	TGTGTTGGGTGTGTTTGG	Lib5_phiX.R1.fastq	Lib5_phiX.R2.fastq	lib5
ds2016-58	GGCCAGAA	TGTGTTGGGTGTGTTTGG	Lib6_phiX.R1.fastq	Lib6_phiX.R2.fastq	lib6
ds2016-10	GGTACTCC	TGTGTTGGGTGTGTTTGG	Lib6_phiX.R1.fastq	Lib6_phiX.R2.fastq	lib6
ds2016-11	TCGGATGC	TGTGTTGGGTGTGTTTGG	Lib7_phiX.R1.fastq	Lib7_phiX.R2.fastq	lib7
ds2015-149	TCTATGAC	TGTGTTGGGTGTGTTTGG	Lib7_phiX.R1.fastq	Lib7_phiX.R2.fastq	lib7
ds2015-162	TTCTGGCT	TGTGTTGGGTGTGTTTGG	Lib8_phiX.R1.fastq	Lib8_phiX.R2.fastq	lib8
ds2015-170	TTGCGTCA	TGTGTTGGGTGTGTTTGG	Lib8_phiX.R1.fastq	Lib8_phiX.R2.fastq	lib8
</pre></div>
</div>
<p>You can add categories for each sample so they can be used when coloring sequences in trees from the Rps2tree module.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Parameters files</a><ul>
<li><a class="reference internal" href="#parameters-yaml">parameters.yaml</a></li>
<li><a class="reference internal" href="#step-yaml">step.yaml</a></li>
<li><a class="reference internal" href="#map-txt">map.txt</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="prerequisite.html"
                        title="previous chapter">Prerequisite</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="modules.html"
                        title="next chapter">Modules description</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/parameter.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="modules.html" title="Modules description"
             >next</a> |</li>
        <li class="right" >
          <a href="prerequisite.html" title="Prerequisite"
             >previous</a> |</li>
        <li><a href="index.html">VirAnnot 0.0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, Sebastien Theil.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3b1.
    </div>
  </body>
</html>