
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Prerequisite &#8212; VirAnnot 0.0.1 documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Parameters files" href="parameter.html" />
    <link rel="prev" title="Welcome to VirAnnot’s documentation!" href="index.html" /> 
  </head>
  <body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="parameter.html" title="Parameters files"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to VirAnnot’s documentation!"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">VirAnnot 0.0.1 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="prerequisite">
<h1>Prerequisite<a class="headerlink" href="#prerequisite" title="Permalink to this headline">¶</a></h1>
<div class="section" id="external-programs">
<h2>External programs<a class="headerlink" href="#external-programs" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>NCBI Blast+ suite (<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST">ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST</a>)</li>
<li>SQLite (<a class="reference external" href="https://www.sqlite.org/">https://www.sqlite.org/</a>)</li>
<li>Mummer3 (<a class="reference external" href="http://mummer.sourceforge.net/">http://mummer.sourceforge.net/</a>)</li>
<li>Bowtie2 (<a class="reference external" href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">http://bowtie-bio.sourceforge.net/bowtie2/index.shtml</a>)</li>
<li>Cutadapt (<a class="reference external" href="https://github.com/marcelm/cutadapt">https://github.com/marcelm/cutadapt</a>)</li>
<li>ETE tree (<a class="reference external" href="http://etetoolkit.org/">http://etetoolkit.org/</a>)</li>
<li>IDBA-UD (<a class="reference external" href="https://github.com/loneknightpy/idba">https://github.com/loneknightpy/idba</a>)</li>
<li>drVM (<a class="reference external" href="https://sourceforge.net/projects/sb2nhri/files/drVM/">https://sourceforge.net/projects/sb2nhri/files/drVM/</a>)</li>
<li>Open Grid Scheduler (<a class="reference external" href="http://gridscheduler.sourceforge.net/">http://gridscheduler.sourceforge.net/</a>)</li>
<li>Diamond (<a class="reference external" href="https://github.com/bbuchfink/diamond">https://github.com/bbuchfink/diamond</a>)</li>
</ul>
</div>
<div class="section" id="external-databases">
<h2>External databases<a class="headerlink" href="#external-databases" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>NCBI nr, nt (<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/blast/db/">ftp://ftp.ncbi.nlm.nih.gov/blast/db/</a>)</li>
<li>NCBI Taxonomy (<a class="reference external" href="ftp://ftp.ncbi.nih.gov/pub/taxonomy">ftp://ftp.ncbi.nih.gov/pub/taxonomy</a>)</li>
<li>PFAM (<a class="reference external" href="ftp://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/">ftp://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/</a>) (rpsblast files, fasta files, and smp files)</li>
</ul>
</div>
<div class="section" id="perl-external-libraries">
<h2>Perl external libraries<a class="headerlink" href="#perl-external-libraries" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Getopt::Long</li>
<li>File::Basename</li>
<li>DBI</li>
<li>Data::Dumper</li>
<li>Bioperl</li>
<li>Color::Rgb</li>
<li>List::Util</li>
<li>Excel::Writer</li>
<li>Log::Log4perl</li>
<li>DBD::SQLite</li>
<li>SQL::SplitStatement</li>
<li>Math::Round</li>
</ul>
</div>
<div class="section" id="perl-included-libraries">
<h2>Perl included libraries<a class="headerlink" href="#perl-included-libraries" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Tools::Fasta</li>
<li>Tools::Fastq</li>
<li>Tools::Blast</li>
<li>Tools::Taxonomy</li>
<li>Logger::Logger</li>
</ul>
</div>
<div class="section" id="python-library">
<h2>Python library<a class="headerlink" href="#python-library" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>os</li>
<li>call</li>
<li>logging</li>
<li>random</li>
<li>string</li>
<li>argparse</li>
<li>re</li>
<li>sys</li>
<li>Bio</li>
<li>time</li>
<li>glob</li>
<li>shutil</li>
<li>yaml</li>
<li>csv</li>
<li>importlib</li>
<li>matplotlib</li>
</ul>
</div>
</div>
<div class="section" id="install">
<h1>Install<a class="headerlink" href="#install" title="Permalink to this headline">¶</a></h1>
<p>Add tools and launchers folders to your $PATH.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="nb">export</span> <span class="nv">PATH</span><span class="o">=</span>/path/to/tools:/path/to/launchers:<span class="nv">$PATH</span>
</pre></div>
</div>
<p>Add lib folder to your $PERL5LIB.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="nb">export</span> <span class="nv">PERL5LIB</span><span class="o">=</span>/path/to/lib:<span class="nv">$PERL5LIB</span>
</pre></div>
</div>
</div>
<div class="section" id="database">
<h1>Database<a class="headerlink" href="#database" title="Permalink to this headline">¶</a></h1>
<p>NCBI taxonomy and the homemade per domain Pfam taxonomy are stored in a simple SQLite database.</p>
<p>Schema:</p>
<img alt="_images/taxo_sql.svg" src="_images/taxo_sql.svg" /><div class="section" id="ncbi-taxonomy">
<h2>NCBI Taxonomy<a class="headerlink" href="#ncbi-taxonomy" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Download and extract NCBI taxonomy files.</li>
</ul>
<div class="highlight-bash"><div class="highlight"><pre><span></span>wget ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz <span class="p">;</span> gunzip taxdump.tar.gz<span class="p">;</span> tar -xf taxdump.tar<span class="p">;</span>
wget ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/prot.accession2taxid.gz <span class="p">;</span> gunzip prot.accession2taxid.gz<span class="p">;</span>
wget ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/nucl_gb.accession2taxid.gz <span class="p">;</span> gunzip nucl_gb.accession2taxid.gz<span class="p">;</span>
</pre></div>
</div>
<p>Optionally you can combine multiple accession2taxid file with a simple cat. But keep separated nucl and prot accessions as they will be loaded in two different tables.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>wget ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/dead_prot.accession2taxid.gz <span class="p">;</span> gunzip dead_prot.accession2taxid.gz<span class="p">;</span>
cat prot.accession2taxid dead_prot.accession2taxid &gt; acc2taxid.prot

wget ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/nucl_wgs.accession2taxid.gz <span class="p">;</span> gunzip nucl_wgs.accession2taxid.gz<span class="p">;</span>
wget ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/dead_wgs.accession2taxid.gz <span class="p">;</span> gunzip dead_wgs.accession2taxid.gz
cat nucl_wgs.accession2taxid nucl_gb.accession2taxid dead_wgs.accession2taxid &gt; acc2taxid.nucl
</pre></div>
</div>
<p>Launch the loadTaxonomy.pl script that will create the sqlite database. The script needs two provided sqlite files: <code class="docutils literal"><span class="pre">taxonomyIndex.sql</span></code> and <code class="docutils literal"><span class="pre">taxonomyStructure.sql</span></code> that describe the database struture.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>loadTaxonomy.pl -struct taxonomyStructure.sql -index taxonomyIndex.sql -acc_prot acc2taxid.prot -acc_nucl acc2taxid.nucl -names names.dmp -nodes nodes.dmp
</pre></div>
</div>
</div>
<div class="section" id="pfam-taxonomy">
<h2>PFAM taxonomy<a class="headerlink" href="#pfam-taxonomy" title="Permalink to this headline">¶</a></h2>
<p>The pipeline modules <code class="docutils literal"><span class="pre">rps2ecsv</span></code> and <code class="docutils literal"><span class="pre">rps2tree</span></code> need taxonomic information of the PFAM domains to work.
You need to extract these informations and load it into the sqlite database.</p>
<ul class="simple">
<li>Extract taxonomic information for each sequence of each PFAM domain and store it in <code class="docutils literal"><span class="pre">*.tax.txt</span></code> files:</li>
</ul>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="se">\l</span>s -1 *.FASTA <span class="p">|</span> sed <span class="s1">&#39;s,^\(.*\)\.FASTA,gi2taxonomy.pl -i &amp; -o \1.tax.txt -r,&#39;</span> <span class="p">|</span> bash
</pre></div>
</div>
<ul class="simple">
<li>Create a file of file for the <code class="docutils literal"><span class="pre">*.tax.txt</span></code> files:</li>
</ul>
<div class="highlight-bash"><div class="highlight"><pre><span></span>listPath.pl -d . <span class="p">|</span> grep <span class="s1">&#39;tax.txt&#39;</span> &gt; idx
</pre></div>
</div>
<ul class="simple">
<li>Compute taxonomy statistic for each domain and create a sql file to load into the database:</li>
</ul>
<div class="highlight-bash"><div class="highlight"><pre><span></span>taxo_profile_to_sql.pl -i idx -o taxo_profile.sql
</pre></div>
</div>
<ul class="simple">
<li>Load into the database:</li>
</ul>
<div class="highlight-bash"><div class="highlight"><pre><span></span>sqlite3 taxonomy.tmp.sqlite &lt; taxo_profile.sql
</pre></div>
</div>
</div>
<div class="section" id="ncbi-blast-database">
<h2>NCBI Blast database<a class="headerlink" href="#ncbi-blast-database" title="Permalink to this headline">¶</a></h2>
<p>NCBI non redundant databases are very large and similarity search using Blast is an intensive task. I recommand to use those databases on computer clusters.</p>
<ul class="simple">
<li>Download NCBI nr et nt Blast files.</li>
</ul>
<div class="highlight-bash"><div class="highlight"><pre><span></span>wget ftp://ftp.ncbi.nlm.nih.gov/blast/db/nr.*.tar.gz
wget ftp://ftp.ncbi.nlm.nih.gov/blast/db/nt.*.tar.gz
</pre></div>
</div>
<p>Modify the parameters.yaml to fit your configuration.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">servers</span><span class="p">:</span>
<span class="n">genotoul</span><span class="p">:</span>
  <span class="n">adress</span><span class="p">:</span> <span class="s1">&#39;genotoul.toulouse.inra.fr&#39;</span>
  <span class="n">username</span><span class="p">:</span> <span class="s1">&#39;stheil&#39;</span>
  <span class="n">db</span><span class="p">:</span>
    <span class="n">nr</span><span class="p">:</span> <span class="s1">&#39;/bank/blastdb/nr&#39;</span>
    <span class="n">nt</span><span class="p">:</span> <span class="s1">&#39;/bank/blastdb/nt&#39;</span>
</pre></div>
</div>
<p>Reduced databases are a good choice for limited computer ressources and drastically faster similarity search. Here are some example commands using NCBI tools to download sequences.</p>
<ul class="simple">
<li>Reduced NCBI databases:</li>
</ul>
<p>Get all viroids nucleotide sequence from genbank:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">esearch</span> <span class="o">-</span><span class="n">db</span> <span class="s2">&quot;nucleotide&quot;</span> <span class="o">-</span><span class="n">query</span> <span class="s2">&quot;txid12884[Organism]&quot;</span> <span class="o">|</span> <span class="n">efetch</span> <span class="o">-</span><span class="nb">format</span> <span class="n">fasta</span> <span class="o">&gt;</span> <span class="n">viroids_nucl</span><span class="o">.</span><span class="n">fna</span>
</pre></div>
</div>
<p>Get all viruses nucleotide sequences from genbank:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">esearch</span> <span class="o">-</span><span class="n">db</span> <span class="s2">&quot;nucleotide&quot;</span> <span class="o">-</span><span class="n">query</span> <span class="s2">&quot;txid10239[Organism]&quot;</span> <span class="o">|</span> <span class="n">efetch</span> <span class="o">-</span><span class="nb">format</span> <span class="n">fasta</span> <span class="o">&gt;</span> <span class="n">viruses_nucl</span><span class="o">.</span><span class="n">fna</span>
</pre></div>
</div>
<p>Create Blast DB example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">makeblastdb</span> <span class="o">-</span><span class="ow">in</span> <span class="n">viruses_nucl</span><span class="o">.</span><span class="n">fna</span> <span class="o">-</span><span class="n">parse_seqids</span> <span class="o">-</span><span class="n">dbtype</span> <span class="n">nucl</span>
</pre></div>
</div>
<ul class="simple">
<li>Download PFAM files for RPSBLAST.</li>
</ul>
<div class="highlight-bash"><div class="highlight"><pre><span></span>wget ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/little_endian/Pfam_LE.tar.gz
wget ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/fasta.tar.gz
wget ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/cdd.tar.gz
</pre></div>
</div>
<p>Here I use only PFAM domains but <code class="docutils literal"><span class="pre">fasta.tar.gz</span></code> and <code class="docutils literal"><span class="pre">cdd.tar.gz</span></code> contains files for the entire CDD database. You can either delete files that are not from PFAM database or use the complete CDD.</p>
<ul class="simple">
<li>Delete file that are not from PFAM:</li>
</ul>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="se">\l</span>s -1 <span class="p">|</span> grep -v <span class="s1">&#39;pfam&#39;</span> <span class="p">|</span> sed <span class="s1">&#39;s,^.*$,rm &amp;,&#39;</span>
</pre></div>
</div>
<p>Add ‘| bash’ if correct.</p>
<ul class="simple">
<li>Download entire CDD database:</li>
</ul>
<div class="highlight-bash"><div class="highlight"><pre><span></span>wget ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/little_endian/CDD_LE.tar.gz
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Prerequisite</a><ul>
<li><a class="reference internal" href="#external-programs">External programs</a></li>
<li><a class="reference internal" href="#external-databases">External databases</a></li>
<li><a class="reference internal" href="#perl-external-libraries">Perl external libraries</a></li>
<li><a class="reference internal" href="#perl-included-libraries">Perl included libraries</a></li>
<li><a class="reference internal" href="#python-library">Python library</a></li>
</ul>
</li>
<li><a class="reference internal" href="#install">Install</a></li>
<li><a class="reference internal" href="#database">Database</a><ul>
<li><a class="reference internal" href="#ncbi-taxonomy">NCBI Taxonomy</a></li>
<li><a class="reference internal" href="#pfam-taxonomy">PFAM taxonomy</a></li>
<li><a class="reference internal" href="#ncbi-blast-database">NCBI Blast database</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Welcome to VirAnnot’s documentation!</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="parameter.html"
                        title="next chapter">Parameters files</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/prerequisite.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="parameter.html" title="Parameters files"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to VirAnnot’s documentation!"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">VirAnnot 0.0.1 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2016, Sebastien Theil.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.3.
    </div>
  </body>
</html>